# Association tests

These exercises introduce association testing: how to find which genetic variants are associated with a phenotype.

## Null model

The first step in an association test is to fit the null model. We will need an `AnnotatedDataFrame` with phenotypes. 
This data structure is provided by the Bioconductor `Biobase` package, and it contains both the data and metadata (descriptions of each column in the data).

```{r}
# sample annotation
sampfile <- "data/sample_phenotype_annotation.RData"
annot <- TopmedPipeline::getobj(sampfile)
library(Biobase)
# access the data with the pData() function
head(pData(annot))
# access the metadata with the varMetadata() function
varMetadata(annot)
```

We will test for an association between genotype and height, adjusting for sex, age, and study as covariates. First, examine the data to see if the height distribution varies by study.

```{r height_variance}
library(ggplot2)
ggplot(pData(annot), aes(study, height)) + geom_boxplot()
```

From the boxplot, it is clear that the different studies have different mean and variance for height. However, it is possible that this could be the result of different sex and age distributions in each study. To check this, we run a model that adjusts height for age and sex.

```{r}
library(GENESIS)
mod_1 <- fitNullModel(annot, outcome="height", covars=c("sex", "age"), 
                      verbose=FALSE)
```

The output of `fitNullModel` is a list with a number of named elements
```{r}
names(mod_1)
```

The elements that we will work with in this exercise are:

* `converged`: an indicator of whether the model successfully converged
* `model.matrix`: The matrix of subject-covariate values used to fit the model
* `fixef`: The fitted fixed effects
* `betaCov`: The covariance of the fitted fixed effects
* `resid.marginal`: The (marginal) residuals from the model, which have been adjusted for the fixed effects but not for the covariance structure
* `varComp`: The fitted variance component for each input covariance matrix

Make sure the model converged.

```{r}
mod_1$converged
```

Now, add the residuals to the phenotype data frame for plotting.

```{r}
annot$residuals <- mod_1$resid.marginal
```

We want to check if the different studies have the same mean height after adjustment for other covariates (here, age and sex).
We will first do this qualitatively by making a boxplot of the residuals by study.
```{r resid_1}
ggplot(pData(annot), aes(study, residuals)) + geom_boxplot()
```

Height still has different variances in each study, even after adjusting for age and sex Therefore, we allow the model to use heterogeneous variance among studies with the parameter `group.var`.

```{r null_model_fit}
nullmod <- fitNullModel(annot, outcome="height", covars=c("sex", "age", "study"), 
                        group.var="study", verbose=FALSE)
save(nullmod, file="null_model.RData")
```

The `fixef` element now includes effects for study:
```{r}
nullmod$fixef
```

We can also check the variance components (`varComp`) in the model, which are different for each study:
```{r}
nullmod$varComp
```

The fitted values of the variance components are different for the different studies, indicating that the distributions of height in the three studies have different variance even after accounting for the other covariates.


We also recommend taking an inverse normal transform of the residuals and refitting the model. This is done separately for each group, and the transformed residuals are rescaled. See the full procedure in the  
[pipeline documenation](https://github.com/UW-GAC/analysis_pipeline#association-testing).

## Single-variant tests

Now that we have a null model adjusting height for covariates, we can run an association test to look for genetic effects on height.

Single-variant tests are the same as in GWAS. We use the `assocTestSingle` function in GENESIS. First, we have to create a `SeqVarData` object including both the GDS file and the sample annotation containing phenotypes. We then create a `SeqVarBlockIterator` object to iterate over blocks of variants.

```{r assoc_single}
library(SeqVarTools)
gdsfile <- "data/1KG_phase3_subset_chr1.gds"
gds <- seqOpen(gdsfile)
seqData <- SeqVarData(gds, sampleData=annot)
iterator <- SeqVarBlockIterator(seqData, verbose=FALSE)
assoc <- assocTestSingle(iterator, nullmod)
head(assoc)
```

We make a QQ plot to examine the results.

```{r assoc_single_qq}
library(ggplot2)
qqPlot <- function(pval) {
    pval <- pval[!is.na(pval)]
    n <- length(pval)
    x <- 1:n
    dat <- data.frame(obs=sort(pval),
                      exp=x/n,
                      upper=qbeta(0.025, x, rev(x)),
                      lower=qbeta(0.975, x, rev(x)))
    
    ggplot(dat, aes(-log10(exp), -log10(obs))) +
        geom_line(aes(-log10(exp), -log10(upper)), color="gray") +
        geom_line(aes(-log10(exp), -log10(lower)), color="gray") +
        geom_point() +
        geom_abline(intercept=0, slope=1, color="red") +
        xlab(expression(paste(-log[10], "(expected P)"))) +
        ylab(expression(paste(-log[10], "(observed P)"))) +
        theme_bw()
}    

qqPlot(assoc$Score.pval)
```

## Exercises

1. Logistic regression: `fitNullModel` can use a binary phenotype as the outcome variable by specifying the argument `family=binomial`. Use the `status` column in the sample annotation to fit a null model for simulated case/control status, with `sex` and `Population` as covariates. Then run a single-variant test using this model.

2. Inverse normal transform: use the function `nullModelInvNorm` to perform an inverse normal transform on the `height` variable. For each study separately, compute a null model and do the inverse normal transform using just the values for that study. Compare these residuals with the initial residuals you obtained for that study by transforming all studies together.


## Sliding window tests

For rare variants, we can do burden tests or SKAT using the GENESIS function `assocTestAggregate`. We restrict the test to variants with alternate allele frequency < 0.1. (For real data, this threshold would be lower.) We use a flat weighting scheme. We define a sliding window across the genome using a `SeqVarWindowIterator`.

```{r assoc_window_burden}
seqResetFilter(seqData, verbose=FALSE)
iterator <- SeqVarWindowIterator(seqData, windowSize=5000, windowShift=2000, verbose=FALSE)
assoc <- assocTestAggregate(iterator, nullmod, test="Burden", AF.max=0.1, weight.beta=c(1,1))
names(assoc)
head(assoc$results)
head(assoc$variantInfo)

qqPlot(assoc$results$Score.pval)
```

For SKAT, we use the Wu weights.

```{r assoc_window_skat}
seqResetFilter(seqData, verbose=FALSE)
iterator <- SeqVarWindowIterator(seqData, windowSize=5000, windowShift=2000, verbose=FALSE)
assoc <- assocTestAggregate(iterator, nullmod, test="SKAT", AF.max=0.1, weight.beta=c(1,25))
head(assoc$results)
head(assoc$variantInfo)

qqPlot(assoc$results$pval_0)
```

## Exercise

3. Repeat the previous exercise on logistic regression, this time running a sliding-window test.


# Association tests - Solutions

1. Logistic regression: `fitNullModel` can use a binary phenotype as the outcome variable by specifying the argument `family=binomial`. Use the `status` column in the sample annotation to fit a null model for simulated case/control status, with `sex` and `Population` as covariates. Then run a single-variant test using this model.

```{r exercise_logistic}
nullmod.status <- fitNullModel(annot, outcome="status", covars=c("sex", "Population"), 
                               family=binomial, verbose=FALSE)
resetIterator(iterator, verbose=FALSE)
assoc <- assocTestSingle(iterator, nullmod.status, test="Score")
head(assoc)
```

2. Inverse normal transform: use the function `nullModelInvNorm` to perform an inverse normal transform on the `height` variable. For each study separately, compute a null model and do the inverse normal transform using just the values for that study. Compare these residuals with the initial residuals you obtained for that study by transforming all studies together.

```{r exercise_invnorm}
nullmod.norm.all <- nullModelInvNorm(nullmod, norm.option="all")
dat.all <- data.frame(sample.id=nullmod.norm.all$sample.id, 
                      resid.norm=nullmod.norm.all$resid.marginal,
                      study=annot$study,
                      run="combined")

nullmod.norm.group <- nullModelInvNorm(nullmod, norm.option="by.group")
dat.group <- data.frame(sample.id=nullmod.norm.group$sample.id, 
                        resid.norm=nullmod.norm.group$resid.marginal,
                        study=annot$study,
                        run="separate")

dat <- rbind(dat.all, dat.group)

ggplot(dat, aes(study, resid.norm, fill=run)) + geom_boxplot()
```


3. Repeat the previous exercise on logistic regression, this time running a sliding-window test.

```{r exercise_sliding}
nullmod.status <- fitNullModel(annot, outcome="status", covars=c("sex", "Population"), 
                               family=binomial, verbose=FALSE)
seqResetFilter(seqData, verbose=FALSE)
iterator <- SeqVarWindowIterator(seqData, windowSize=5000, windowShift=2000, verbose=FALSE)
assoc <- assocTestAggregate(iterator, nullmod, test="SKAT", AF.max=0.1, weight.beta=c(1,25))
head(assoc$results)
```

```{r assoc_close}
seqClose(gds)
```