---
title: "WISG Module 4: Computational Pipeline for WGS Data"
date: "`r Sys.Date()`"
site: bookdown::bookdown_site
output: bookdown::gitbook
documentclass: book
link-citations: yes
github-repo: UW-GAC/WISG_2019
description: "Course materials for WISG Module 4"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE, message=FALSE)
```

# Introduction

This site contains course materials for WISG Module 4: Computational Pipeline for WGS Data, January 23-24, 2019.
Data used is located in the [github repository](https://github.com/UW-GAC/WISG_2019) from which the site is built, as well as in the [TOPMed analysis pipeline](https://github.com/UW-GAC/analysis_pipeline).

## Schedule

[Lecture slides](https://drive.google.com/drive/folders/1-sp3l6qumvFjohAqMy8WC513PXeHqeiJ?usp=sharing)

**Wednesday, January 23**

- Introduction
- Sequencing and data formats
    - Sequencing overview
    - Sequencing data formats
    - Intro to Genomic Data Storage
- Association tests
    - Methods and motivation
    - GENESIS for association tests


**Thursday, January 24**

- Population structure and relatedness
    - Population structure inference
    - Relatedness inference
    - R packages for PCA and relatedness
    - Using GENESIS with mixed models
- Variant annotation
- Pipeline design and examples
    - Cloud computing introduction
    - Analysis pipeline design
    - Analysis pipeline on the cloud
- Cloud platforms


## Software requirements

- [R 3.5.1](https://cran.r-project.org/)
- R packages
    - [SeqArray](http://bioconductor.org/packages/release/bioc/html/SeqArray.html)
    - [SeqVarTools](http://bioconductor.org/packages/release/bioc/html/SeqVarTools.html)
    - [GENESIS](http://bioconductor.org/packages/release/bioc/html/GENESIS.html)
    - [SNPRelate](http://bioconductor.org/packages/release/bioc/html/SNPRelate.html)
    - [TopmedPipeline](https://github.com/UW-GAC/analysis_pipeline/tree/master/TopmedPipeline)
    - dplyr
    - ggplot2
    - GGally
    - survey
    - CompQuadForm
    - TxDb.Hsapiens.UCSC.hg19.knownGene
    
Installation instructions

```{r install, eval=FALSE}
install.packages(c("BiocManager", "remotes"), repos="https://cloud.r-project.org")

BiocManager::install(c("SeqArray", "SeqVarTools", "SNPRelate", "GENESIS", 
                       "survey", "CompQuadForm", "dplyr", "ggplot2", "GGally",
                       "TxDb.Hsapiens.UCSC.hg19.knownGene"))

remotes::install_github("UW-GAC/analysis_pipeline/TopmedPipeline")
```

## Resources

If you are new to R, you might find the following material helpful:

- [Graphics with ggplot2](http://tutorials.iq.harvard.edu/R/Rgraphics/Rgraphics.html) tutorial
- Data manipulation with [dplyr](http://dplyr.tidyverse.org/)
