<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>SISG Module 12: Computational Pipeline for WGS Data</title>
  <meta name="description" content="Course materials for SISG Module 12">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="SISG Module 12: Computational Pipeline for WGS Data" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Course materials for SISG Module 12" />
  <meta name="github-repo" content="UW-GAC/topmed_workshop_2018" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="SISG Module 12: Computational Pipeline for WGS Data" />
  
  <meta name="twitter:description" content="Course materials for SISG Module 12" />
  



<meta name="date" content="2018-07-22">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="association-tests.html">
<link rel="next" href="using-docker.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#schedule"><i class="fa fa-check"></i><b>1.1</b> Schedule</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#resources"><i class="fa fa-check"></i><b>1.2</b> Resources</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="gds-format.html"><a href="gds-format.html"><i class="fa fa-check"></i><b>2</b> GDS format</a><ul>
<li class="chapter" data-level="2.1" data-path="gds-format.html"><a href="gds-format.html#exploring-a-gds-file"><i class="fa fa-check"></i><b>2.1</b> Exploring a GDS file</a></li>
<li class="chapter" data-level="2.2" data-path="gds-format.html"><a href="gds-format.html#exercises"><i class="fa fa-check"></i><b>2.2</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="computing-a-grm.html"><a href="computing-a-grm.html"><i class="fa fa-check"></i><b>3</b> Computing a GRM</a></li>
<li class="chapter" data-level="4" data-path="pc-relate.html"><a href="pc-relate.html"><i class="fa fa-check"></i><b>4</b> PC-Relate</a><ul>
<li class="chapter" data-level="4.1" data-path="pc-relate.html"><a href="pc-relate.html#king"><i class="fa fa-check"></i><b>4.1</b> KING</a></li>
<li class="chapter" data-level="4.2" data-path="pc-relate.html"><a href="pc-relate.html#pc-air"><i class="fa fa-check"></i><b>4.2</b> PC-AiR</a></li>
<li class="chapter" data-level="4.3" data-path="pc-relate.html"><a href="pc-relate.html#pc-relate-1"><i class="fa fa-check"></i><b>4.3</b> PC-Relate</a></li>
<li class="chapter" data-level="4.4" data-path="pc-relate.html"><a href="pc-relate.html#comparison-with-pedigree"><i class="fa fa-check"></i><b>4.4</b> Comparison with pedigree</a></li>
<li class="chapter" data-level="4.5" data-path="pc-relate.html"><a href="pc-relate.html#exercise"><i class="fa fa-check"></i><b>4.5</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html"><i class="fa fa-check"></i><b>5</b> Phenotype Harmonization</a><ul>
<li class="chapter" data-level="5.1" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#inspect-individual-study-data-in-r"><i class="fa fa-check"></i><b>5.1</b> Inspect individual study data in R</a></li>
<li class="chapter" data-level="5.2" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#compare-study-values"><i class="fa fa-check"></i><b>5.2</b> Compare study values</a></li>
<li class="chapter" data-level="5.3" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#using-mixed-models-to-compare-studies"><i class="fa fa-check"></i><b>5.3</b> Using mixed models to compare studies</a><ul>
<li class="chapter" data-level="5.3.1" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#create-an-annotated-data-frame"><i class="fa fa-check"></i><b>5.3.1</b> Create an Annotated Data Frame</a></li>
<li class="chapter" data-level="5.3.2" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#obtain-the-genetic-relatedness-matrix"><i class="fa fa-check"></i><b>5.3.2</b> Obtain the genetic relatedness matrix</a></li>
<li class="chapter" data-level="5.3.3" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#fit-a-mixed-model-without-study"><i class="fa fa-check"></i><b>5.3.3</b> Fit a mixed model without study</a></li>
<li class="chapter" data-level="5.3.4" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#fit-a-model-with-study"><i class="fa fa-check"></i><b>5.3.4</b> Fit a model with study</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#final-considerations"><i class="fa fa-check"></i><b>5.4</b> Final considerations</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="association-tests.html"><a href="association-tests.html"><i class="fa fa-check"></i><b>6</b> Association tests</a><ul>
<li class="chapter" data-level="6.1" data-path="association-tests.html"><a href="association-tests.html#null-model"><i class="fa fa-check"></i><b>6.1</b> Null model</a></li>
<li class="chapter" data-level="6.2" data-path="association-tests.html"><a href="association-tests.html#single-variant-tests"><i class="fa fa-check"></i><b>6.2</b> Single-variant tests</a></li>
<li class="chapter" data-level="6.3" data-path="association-tests.html"><a href="association-tests.html#exercises-1"><i class="fa fa-check"></i><b>6.3</b> Exercises</a></li>
<li class="chapter" data-level="6.4" data-path="association-tests.html"><a href="association-tests.html#sliding-window-tests"><i class="fa fa-check"></i><b>6.4</b> Sliding window tests</a><ul>
<li class="chapter" data-level="6.4.1" data-path="association-tests.html"><a href="association-tests.html#exercise-1"><i class="fa fa-check"></i><b>6.4.1</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="association-tests.html"><a href="association-tests.html#annotation-based-aggregate-tests"><i class="fa fa-check"></i><b>6.5</b> Annotation-based aggregate tests</a><ul>
<li class="chapter" data-level="6.5.1" data-path="association-tests.html"><a href="association-tests.html#working-with-variant-annotation"><i class="fa fa-check"></i><b>6.5.1</b> Working with variant annotation</a></li>
<li class="chapter" data-level="6.5.2" data-path="association-tests.html"><a href="association-tests.html#aggregating-and-filtering-variants-using-annotation"><i class="fa fa-check"></i><b>6.5.2</b> Aggregating and filtering variants using annotation</a></li>
<li class="chapter" data-level="6.5.3" data-path="association-tests.html"><a href="association-tests.html#aggregate-unit-for-association-testing-exercise"><i class="fa fa-check"></i><b>6.5.3</b> Aggregate unit for association testing exercise</a></li>
<li class="chapter" data-level="6.5.4" data-path="association-tests.html"><a href="association-tests.html#association-testing-with-aggregate-units"><i class="fa fa-check"></i><b>6.5.4</b> Association testing with aggregate units</a></li>
<li class="chapter" data-level="6.5.5" data-path="association-tests.html"><a href="association-tests.html#exercise-2"><i class="fa fa-check"></i><b>6.5.5</b> Exercise</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html"><i class="fa fa-check"></i><b>7</b> Analysis Pipeline</a><ul>
<li class="chapter" data-level="7.1" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#running-on-a-local-cluster"><i class="fa fa-check"></i><b>7.1</b> Running on a local cluster</a></li>
<li class="chapter" data-level="7.2" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#running-on-aws-batch"><i class="fa fa-check"></i><b>7.2</b> Running on AWS Batch</a><ul>
<li class="chapter" data-level="7.2.1" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#log-into-aws-docker-image"><i class="fa fa-check"></i><b>7.2.1</b> Log into AWS docker image</a></li>
<li class="chapter" data-level="7.2.2" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#download-the-docker-helper-functions"><i class="fa fa-check"></i><b>7.2.2</b> Download the docker helper functions</a></li>
<li class="chapter" data-level="7.2.3" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#cd-to-working-directory-and-create-config-file"><i class="fa fa-check"></i><b>7.2.3</b> cd to working directory and create config file</a></li>
<li class="chapter" data-level="7.2.4" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#execute-the-python-helper-function-to-run-the-docker-image"><i class="fa fa-check"></i><b>7.2.4</b> Execute the python helper function to run the docker image</a></li>
<li class="chapter" data-level="7.2.5" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#print-out-aws-commands-if-executing-the-pipeline"><i class="fa fa-check"></i><b>7.2.5</b> Print out AWS commands if executing the pipeline</a></li>
<li class="chapter" data-level="7.2.6" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#execute-the-pipeline"><i class="fa fa-check"></i><b>7.2.6</b> Execute the pipeline</a></li>
<li class="chapter" data-level="7.2.7" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#monitor-the-jobs"><i class="fa fa-check"></i><b>7.2.7</b> Monitor the jobs</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="using-docker.html"><a href="using-docker.html"><i class="fa fa-check"></i><b>8</b> Using Docker</a><ul>
<li class="chapter" data-level="8.1" data-path="using-docker.html"><a href="using-docker.html#prerequisites"><i class="fa fa-check"></i><b>8.1</b> Prerequisites</a></li>
<li class="chapter" data-level="8.2" data-path="using-docker.html"><a href="using-docker.html#general-steps"><i class="fa fa-check"></i><b>8.2</b> General Steps</a><ul>
<li class="chapter" data-level="8.2.1" data-path="using-docker.html"><a href="using-docker.html#download-the-docker-image"><i class="fa fa-check"></i><b>8.2.1</b> Download the Docker Image</a></li>
<li class="chapter" data-level="8.2.2" data-path="using-docker.html"><a href="using-docker.html#create-working-directory"><i class="fa fa-check"></i><b>8.2.2</b> Create Working Directory</a></li>
<li class="chapter" data-level="8.2.3" data-path="using-docker.html"><a href="using-docker.html#download-the-docker-helper-functions-1"><i class="fa fa-check"></i><b>8.2.3</b> Download the docker helper functions</a></li>
<li class="chapter" data-level="8.2.4" data-path="using-docker.html"><a href="using-docker.html#running-rstudio"><i class="fa fa-check"></i><b>8.2.4</b> Running RStudio</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="using-docker.html"><a href="using-docker.html#examples"><i class="fa fa-check"></i><b>8.3</b> Examples</a><ul>
<li class="chapter" data-level="8.3.1" data-path="using-docker.html"><a href="using-docker.html#example-1"><i class="fa fa-check"></i><b>8.3.1</b> Example 1</a></li>
<li class="chapter" data-level="8.3.2" data-path="using-docker.html"><a href="using-docker.html#example-2"><i class="fa fa-check"></i><b>8.3.2</b> Example 2</a></li>
<li class="chapter" data-level="8.3.3" data-path="using-docker.html"><a href="using-docker.html#example-3"><i class="fa fa-check"></i><b>8.3.3</b> Example 3</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="analysis-commons.html"><a href="analysis-commons.html"><i class="fa fa-check"></i><b>9</b> Analysis Commons</a><ul>
<li class="chapter" data-level="9.1" data-path="analysis-commons.html"><a href="analysis-commons.html#outline"><i class="fa fa-check"></i><b>9.1</b> Outline</a></li>
<li class="chapter" data-level="9.2" data-path="analysis-commons.html"><a href="analysis-commons.html#web-interface-and-running-an-analysis-application"><i class="fa fa-check"></i><b>9.2</b> Web Interface and Running an Analysis Application</a><ul>
<li class="chapter" data-level="9.2.1" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-1-run-a-single-variant-analysis."><i class="fa fa-check"></i><b>9.2.1</b> Exercise 1) Run a single variant analysis.</a></li>
<li class="chapter" data-level="9.2.2" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-2-run-skat-test-grouping-variants-into-gene-transcript-regions-and-limit-the-variants-to-those-with-a-cadd-phred-score-2-and-maf-5."><i class="fa fa-check"></i><b>9.2.2</b> Exercise 2) Run SKAT test grouping variants into gene transcript regions and limit the variants to those with a CADD phred score &gt; 2 and MAF &lt;= 5%.</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="analysis-commons.html"><a href="analysis-commons.html#command-line-interface"><i class="fa fa-check"></i><b>9.3</b> Command line interface</a><ul>
<li class="chapter" data-level="9.3.1" data-path="analysis-commons.html"><a href="analysis-commons.html#log-in-to-awi"><i class="fa fa-check"></i><b>9.3.1</b> Log in to AWI</a></li>
<li class="chapter" data-level="9.3.2" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-3-navigate-directories-make-output-directory-examine-files"><i class="fa fa-check"></i><b>9.3.2</b> Exercise 3) Navigate directories, make output directory, examine files</a></li>
<li class="chapter" data-level="9.3.3" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-4-run-single-variant-analysis-from-command-line-using-bash-script"><i class="fa fa-check"></i><b>9.3.3</b> Exercise 4) Run single variant analysis from command line using bash script</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="analysis-commons.html"><a href="analysis-commons.html#writing-your-own-apps"><i class="fa fa-check"></i><b>9.4</b> Writing your own Apps</a><ul>
<li class="chapter" data-level="9.4.1" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-5-write-an-app-that-creates-phenotype-residuals-and-performs-an-inverse-normal-transform"><i class="fa fa-check"></i><b>9.4.1</b> Exercise 5) Write an App that creates phenotype residuals and performs an inverse normal transform</a></li>
<li class="chapter" data-level="9.4.2" data-path="analysis-commons.html"><a href="analysis-commons.html#optional-exercise-6-make-qq-plot"><i class="fa fa-check"></i><b>9.4.2</b> Optional Exercise 6) Make QQ plot</a></li>
<li class="chapter" data-level="9.4.3" data-path="analysis-commons.html"><a href="analysis-commons.html#optional-exercise-8-create-a-regional-association-plot-using-ld-extracted-from-your-data-set"><i class="fa fa-check"></i><b>9.4.3</b> Optional Exercise 8) Create a regional association plot using LD extracted from your data set</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">SISG Module 12: Computational Pipeline for WGS Data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="analysis-pipeline" class="section level1">
<h1><span class="header-section-number">7</span> Analysis Pipeline</h1>
<p>The DCC’s analysis pipeline is hosted on github: <a href="https://github.com/UW-GAC/analysis_pipeline" class="uri">https://github.com/UW-GAC/analysis_pipeline</a></p>
<div id="running-on-a-local-cluster" class="section level2">
<h2><span class="header-section-number">7.1</span> Running on a local cluster</h2>
<p>To run a burden test on our local SGE cluster, first we create a config file and call it <code>assoc_window_burden.config</code>:</p>
<pre><code>out_prefix &quot;test&quot;
gds_file &quot;testdata/1KG_phase3_subset_chr .gds&quot;
phenotype_file &quot;testdata/1KG_phase3_subset_annot.RData&quot;
pcrelate_file &quot;testdata/round2_pcrelate.gds&quot;
pca_file &quot;testdata/round2_pcair.RData&quot;
sample_include_file &quot;testdata/sample_include.RData&quot;
variant_include_file &quot;testdata/variant_include_chr .RData&quot;
outcome outcome
covars &quot;sex&quot;
n_pcs 4
alt_freq_max &quot;0.1&quot;
test &quot;burden&quot;
test_type &quot;score&quot;</code></pre>
<p>We will use the python script <code>assoc.py</code> to submit all jobs. First we look at the available options:</p>
<pre><code>setenv PIPELINE /projects/topmed/working_code/analysis_pipeline_2.0.1
$PIPELINE/assoc.py --help</code></pre>
<p>Let’s run a sliding window test on chromosomes 1-10. We will also specify the cluster type, although UW_Cluster is actually the default. The cluster file is a JSON file that can override default values for the cluster configuration. In this case, we are changing the memory requirements for each job to only reserve a small amount of memory on each cluster node. The last argument is our config file.</p>
<p>First, we print the commands that will be be run without actually submitting jobs:</p>
<pre><code>$PIPELINE/assoc.py \
    --chromosomes 1-10 \
    --cluster_type UW_Cluster \
    --cluster_file test_cluster_cfg.json \
    --print_only \
    window \
    testdata/assoc_window_burden.config</code></pre>
<p>The default segment length is 10,000 kb, but we can change that to 50,000 kb when we submit:</p>
<pre><code>$PIPELINE/assoc.py \
    --chromosomes 1-10 \
    --cluster_type UW_Cluster \
    --cluster_file test_cluster_cfg.json \
    --segment_length 50000 \
    window \
    testdata/assoc_window_burden.config</code></pre>
<p>We can use the <code>qstat</code> command to check the status of our jobs.</p>
</div>
<div id="running-on-aws-batch" class="section level2">
<h2><span class="header-section-number">7.2</span> Running on AWS Batch</h2>
<p>To run a burden test on AWS Batch, we do the following general steps:</p>
<ol style="list-style-type: decimal">
<li>Log into the the docker AMI instance</li>
<li>Download the docker helper functions</li>
<li>cd to a working directory on our NFS volume</li>
<li>Create the configuration file <code>assoc_window_burden.config</code></li>
<li>Execute the python helper function to run the docker image</li>
<li>Optionally execute the association pipeline specifying the AWS Batch service to print out the commands (not running the pipeline)</li>
<li>Execute the association pipeline specifying the AWS Batch service to run the pipeline</li>
<li>Monitor the pipeline via the AWS Batch console</li>
</ol>
<div id="log-into-aws-docker-image" class="section level3">
<h3><span class="header-section-number">7.2.1</span> Log into AWS docker image</h3>
<p>ssh into our image which is running docker. Various docker commands can be executed including running TOPMed version of R (note: TOPMed data is not part of the docker image).</p>
<pre><code>ssh -i ~/.ssh/&lt;some private key&gt; kuraisa@52.27.98.54
[kuraisa@ip-172-255-46-100]~
_4816$ docker images
...
[kuraisa@ip-172-255-46-100]~
_4817$ docker run -it uwgac/r-topmed:dev
/# which R
...
/# R
...
&gt; .libPaths()
...
&gt; library(SeqArray)
...
&gt; q()
...
/# exit
[kuraisa@ip-172-255-46-100]~
_4818$  </code></pre>
</div>
<div id="download-the-docker-helper-functions" class="section level3">
<h3><span class="header-section-number">7.2.2</span> Download the docker helper functions</h3>
<pre><code>git clone https://github.com/uw-gac/docker_helpers
alias pipeline=&#39;/home/kuraisa/docker_helpers/analysis_pipeline.py&#39;</code></pre>
</div>
<div id="cd-to-working-directory-and-create-config-file" class="section level3">
<h3><span class="header-section-number">7.2.3</span> cd to working directory and create config file</h3>
<pre><code>cd /projects/topmed/analysts/kuraisa/workshop/burden
vi assoc_window_burden.config
...</code></pre>
</div>
<div id="execute-the-python-helper-function-to-run-the-docker-image" class="section level3">
<h3><span class="header-section-number">7.2.4</span> Execute the python helper function to run the docker image</h3>
<pre><code>pipeline</code></pre>
</div>
<div id="print-out-aws-commands-if-executing-the-pipeline" class="section level3">
<h3><span class="header-section-number">7.2.5</span> Print out AWS commands if executing the pipeline</h3>
<pre><code>/usr/local/analysis_pipeline/assoc.py \
 single testdata/assoc_window_burden.config \
  --cluster_type AWS_Batch \
  --cluster_file custom_batch.json \
  --print &gt; single_print.log  2&gt;&amp;1</code></pre>
</div>
<div id="execute-the-pipeline" class="section level3">
<h3><span class="header-section-number">7.2.6</span> Execute the pipeline</h3>
<pre><code>/usr/local/analysis_pipeline/assoc.py \
 single testdata/assoc_window_burden.config \
  --cluster_type AWS_Batch \
  --cluster_file custom_batch.json  &gt; single_burden.log  2&gt;&amp;1
</code></pre>
</div>
<div id="monitor-the-jobs" class="section level3">
<h3><span class="header-section-number">7.2.7</span> Monitor the jobs</h3>
<p>From the web browser, log into the AWS account and select the <strong>Batch Services</strong> to monitor:</p>
<ul>
<li>Summary via <strong>Dashboard</strong></li>
<li>Job queue <strong>Optimal_topmed_testdata</strong></li>
<li>View high-level job logs</li>
</ul>
<p>You can switch to <strong>ec2 services</strong> to monitor instances being created or running to support the various jobs.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="association-tests.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="using-docker.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
