<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>4 Association tests | WISG Module 4: Computational Pipeline for WGS Data</title>
  <meta name="description" content="Course materials for WISG Module 4">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="4 Association tests | WISG Module 4: Computational Pipeline for WGS Data />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Course materials for WISG Module 4" />
  <meta name="github-repo" content="UW-GAC/WISG_2019" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="4 Association tests | WISG Module 4: Computational Pipeline for WGS Data />
  
  <meta name="twitter:description" content="Course materials for WISG Module 4" />
  



<meta name="date" content="2019-01-08">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="gds-solutions.html">
<link rel="next" href="association-tests-solutions.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#schedule"><i class="fa fa-check"></i><b>1.1</b> Schedule</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#software-requirements"><i class="fa fa-check"></i><b>1.2</b> Software requirements</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#resources"><i class="fa fa-check"></i><b>1.3</b> Resources</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="gds-format.html"><a href="gds-format.html"><i class="fa fa-check"></i><b>2</b> GDS format</a><ul>
<li class="chapter" data-level="2.1" data-path="gds-format.html"><a href="gds-format.html#exploring-a-gds-file"><i class="fa fa-check"></i><b>2.1</b> Exploring a GDS file</a></li>
<li class="chapter" data-level="2.2" data-path="gds-format.html"><a href="gds-format.html#exercises"><i class="fa fa-check"></i><b>2.2</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="gds-solutions.html"><a href="gds-solutions.html"><i class="fa fa-check"></i><b>3</b> GDS - Solutions</a></li>
<li class="chapter" data-level="4" data-path="association-tests.html"><a href="association-tests.html"><i class="fa fa-check"></i><b>4</b> Association tests</a><ul>
<li class="chapter" data-level="4.1" data-path="association-tests.html"><a href="association-tests.html#null-model"><i class="fa fa-check"></i><b>4.1</b> Null model</a></li>
<li class="chapter" data-level="4.2" data-path="association-tests.html"><a href="association-tests.html#single-variant-tests"><i class="fa fa-check"></i><b>4.2</b> Single-variant tests</a></li>
<li class="chapter" data-level="4.3" data-path="association-tests.html"><a href="association-tests.html#exercises-1"><i class="fa fa-check"></i><b>4.3</b> Exercises</a></li>
<li class="chapter" data-level="4.4" data-path="association-tests.html"><a href="association-tests.html#sliding-window-tests"><i class="fa fa-check"></i><b>4.4</b> Sliding window tests</a></li>
<li class="chapter" data-level="4.5" data-path="association-tests.html"><a href="association-tests.html#exercise"><i class="fa fa-check"></i><b>4.5</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="association-tests-solutions.html"><a href="association-tests-solutions.html"><i class="fa fa-check"></i><b>5</b> Association tests - Solutions</a></li>
<li class="chapter" data-level="6" data-path="computing-a-grm.html"><a href="computing-a-grm.html"><i class="fa fa-check"></i><b>6</b> Computing a GRM</a></li>
<li class="chapter" data-level="7" data-path="pc-relate.html"><a href="pc-relate.html"><i class="fa fa-check"></i><b>7</b> PC-Relate</a><ul>
<li class="chapter" data-level="7.1" data-path="pc-relate.html"><a href="pc-relate.html#king"><i class="fa fa-check"></i><b>7.1</b> KING</a></li>
<li class="chapter" data-level="7.2" data-path="pc-relate.html"><a href="pc-relate.html#pc-air"><i class="fa fa-check"></i><b>7.2</b> PC-AiR</a></li>
<li class="chapter" data-level="7.3" data-path="pc-relate.html"><a href="pc-relate.html#pc-relate-1"><i class="fa fa-check"></i><b>7.3</b> PC-Relate</a></li>
<li class="chapter" data-level="7.4" data-path="pc-relate.html"><a href="pc-relate.html#comparison-with-pedigree"><i class="fa fa-check"></i><b>7.4</b> Comparison with pedigree</a></li>
<li class="chapter" data-level="7.5" data-path="pc-relate.html"><a href="pc-relate.html#exercise-1"><i class="fa fa-check"></i><b>7.5</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="mixed-models.html"><a href="mixed-models.html"><i class="fa fa-check"></i><b>8</b> Mixed models</a><ul>
<li class="chapter" data-level="8.1" data-path="mixed-models.html"><a href="mixed-models.html#null-model-1"><i class="fa fa-check"></i><b>8.1</b> Null model</a></li>
<li class="chapter" data-level="8.2" data-path="mixed-models.html"><a href="mixed-models.html#single-variant-tests-1"><i class="fa fa-check"></i><b>8.2</b> Single-variant tests</a></li>
<li class="chapter" data-level="8.3" data-path="mixed-models.html"><a href="mixed-models.html#exercise-2"><i class="fa fa-check"></i><b>8.3</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="mixed-models-solutions.html"><a href="mixed-models-solutions.html"><i class="fa fa-check"></i><b>9</b> Mixed models - Solutions</a></li>
<li class="chapter" data-level="10" data-path="variant-annotation.html"><a href="variant-annotation.html"><i class="fa fa-check"></i><b>10</b> Variant annotation</a><ul>
<li class="chapter" data-level="10.1" data-path="variant-annotation.html"><a href="variant-annotation.html#using-bioconductor-annotation-resources"><i class="fa fa-check"></i><b>10.1</b> Using Bioconductor annotation resources</a></li>
<li class="chapter" data-level="10.2" data-path="variant-annotation.html"><a href="variant-annotation.html#aggregating-and-filtering-variants-using-annotation"><i class="fa fa-check"></i><b>10.2</b> Aggregating and filtering variants using annotation</a></li>
<li class="chapter" data-level="10.3" data-path="variant-annotation.html"><a href="variant-annotation.html#aggregate-unit-for-association-testing-exercise"><i class="fa fa-check"></i><b>10.3</b> Aggregate unit for association testing exercise</a></li>
<li class="chapter" data-level="10.4" data-path="variant-annotation.html"><a href="variant-annotation.html#association-testing-with-aggregate-units"><i class="fa fa-check"></i><b>10.4</b> Association testing with aggregate units</a></li>
<li class="chapter" data-level="10.5" data-path="variant-annotation.html"><a href="variant-annotation.html#exercise-3"><i class="fa fa-check"></i><b>10.5</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="annotation-solutions.html"><a href="annotation-solutions.html"><i class="fa fa-check"></i><b>11</b> Annotation - Solutions</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">WISG Module 4: Computational Pipeline for WGS Data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="association-tests" class="section level1">
<h1><span class="header-section-number">4</span> Association tests</h1>
<p>These exercises introduce association testing: how to find which genetic variants are associated with a phenotype.</p>
<div id="null-model" class="section level2">
<h2><span class="header-section-number">4.1</span> Null model</h2>
<p>The first step in an association test is to fit the null model. We will need an <code>AnnotatedDataFrame</code> with phenotypes. This data structure is provided by the Bioconductor <code>Biobase</code> package, and it contains both the data and metadata (descriptions of each column in the data).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># sample annotation</span>
sampfile &lt;-<span class="st"> &quot;data/sample_phenotype_annotation.RData&quot;</span>
annot &lt;-<span class="st"> </span>TopmedPipeline<span class="op">::</span><span class="kw">getobj</span>(sampfile)
<span class="kw">library</span>(Biobase)
<span class="co"># access the data with the pData() function</span>
<span class="kw">head</span>(<span class="kw">pData</span>(annot))</code></pre></div>
<pre><code>##   sample.id subject.id Population          Population.Description sex
## 1   HG00096    HG00096        GBR British in England and Scotland   M
## 2   HG00097    HG00097        GBR British in England and Scotland   F
## 3   HG00099    HG00099        GBR British in England and Scotland   F
## 4   HG00100    HG00100        GBR British in England and Scotland   F
## 5   HG00101    HG00101        GBR British in England and Scotland   M
## 6   HG00102    HG00102        GBR British in England and Scotland   F
##   status age  height   study
## 1      0  47 165.300 study_1
## 2      1  47 144.780 study_3
## 3      0  40 185.500 study_2
## 4      1  45 150.622 study_3
## 5      0  40 177.800 study_3
## 6      0  49 169.100 study_1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># access the metadata with the varMetadata() function</span>
<span class="kw">varMetadata</span>(annot)</code></pre></div>
<pre><code>##                                     labelDescription
## sample.id                          sample identifier
## subject.id                        subject identifier
## Population                   population abbreviation
## Population.Description        population description
## sex                                              sex
## status                 simulated case/control status
## age                     age at measurement of height
## height                        subject&#39;s height in cm
## study                               study identifier</code></pre>
<p>We will test for an association between genotype and height, adjusting for sex, age, and study as covariates. First, examine the data to see if the height distribution varies by study.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)
<span class="kw">ggplot</span>(<span class="kw">pData</span>(annot), <span class="kw">aes</span>(study, height)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_boxplot</span>()</code></pre></div>
<p><img src="_main_files/figure-html/height_variance-1.png" width="672" /></p>
<p>From the boxplot, it is clear that the different studies have different mean and variance for height. However, it is possible that this could be the result of different sex and age distributions in each study. To check this, we run a model that adjusts height for age and sex.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(GENESIS)
mod_<span class="dv">1</span> &lt;-<span class="st"> </span><span class="kw">fitNullModel</span>(annot, <span class="dt">outcome=</span><span class="st">&quot;height&quot;</span>, <span class="dt">covars=</span><span class="kw">c</span>(<span class="st">&quot;sex&quot;</span>, <span class="st">&quot;age&quot;</span>), 
                      <span class="dt">verbose=</span><span class="ot">FALSE</span>)</code></pre></div>
<p>The output of <code>fitNullModel</code> is a list with a number of named elements</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(mod_<span class="dv">1</span>)</code></pre></div>
<pre><code>##  [1] &quot;family&quot;         &quot;hetResid&quot;       &quot;varComp&quot;        &quot;varCompCov&quot;    
##  [5] &quot;fixef&quot;          &quot;betaCov&quot;        &quot;fitted.values&quot;  &quot;resid.marginal&quot;
##  [9] &quot;logLik&quot;         &quot;AIC&quot;            &quot;workingY&quot;       &quot;outcome&quot;       
## [13] &quot;model.matrix&quot;   &quot;group.idx&quot;      &quot;cholSigmaInv&quot;   &quot;converged&quot;     
## [17] &quot;zeroFLAG&quot;       &quot;RSS&quot;            &quot;Ytilde&quot;         &quot;resid&quot;         
## [21] &quot;CX&quot;             &quot;CXCXI&quot;          &quot;sample.id&quot;</code></pre>
<p>The elements that we will work with in this exercise are:</p>
<ul>
<li><code>converged</code>: an indicator of whether the model successfully converged</li>
<li><code>model.matrix</code>: The matrix of subject-covariate values used to fit the model</li>
<li><code>fixef</code>: The fitted fixed effects</li>
<li><code>betaCov</code>: The covariance of the fitted fixed effects</li>
<li><code>resid.marginal</code>: The (marginal) residuals from the model, which have been adjusted for the fixed effects but not for the covariance structure</li>
<li><code>varComp</code>: The fitted variance component for each input covariance matrix</li>
</ul>
<p>Make sure the model converged.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mod_<span class="dv">1</span><span class="op">$</span>converged</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>Now, add the residuals to the phenotype data frame for plotting.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">annot<span class="op">$</span>residuals &lt;-<span class="st"> </span>mod_<span class="dv">1</span><span class="op">$</span>resid.marginal</code></pre></div>
<p>We want to check if the different studies have the same mean height after adjustment for other covariates (here, age and sex). We will first do this qualitatively by making a boxplot of the residuals by study.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="kw">pData</span>(annot), <span class="kw">aes</span>(study, residuals)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_boxplot</span>()</code></pre></div>
<p><img src="_main_files/figure-html/resid_1-1.png" width="672" /></p>
<p>Height still has different variances in each study, even after adjusting for age and sex Therefore, we allow the model to use heterogeneous variance among studies with the parameter <code>group.var</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nullmod &lt;-<span class="st"> </span><span class="kw">fitNullModel</span>(annot, <span class="dt">outcome=</span><span class="st">&quot;height&quot;</span>, <span class="dt">covars=</span><span class="kw">c</span>(<span class="st">&quot;sex&quot;</span>, <span class="st">&quot;age&quot;</span>, <span class="st">&quot;study&quot;</span>), 
                        <span class="dt">group.var=</span><span class="st">&quot;study&quot;</span>, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
<span class="kw">save</span>(nullmod, <span class="dt">file=</span><span class="st">&quot;null_model.RData&quot;</span>)</code></pre></div>
<p>The <code>fixef</code> element now includes effects for study:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nullmod<span class="op">$</span>fixef</code></pre></div>
<pre><code>##                       Est         SE        Stat         pval
## (Intercept)  163.67175933 3.18936046 2633.542254 0.000000e+00
## sexM           6.28764510 0.68812251   83.491933 6.397652e-20
## age            0.07519782 0.06921691    1.180283 2.772984e-01
## studystudy_2  10.63152325 0.82176939  167.375183 2.769991e-38
## studystudy_3  -8.96183691 0.84479021  112.537257 2.724960e-26</code></pre>
<p>We can also check the variance components (<code>varComp</code>) in the model, which are different for each study:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nullmod<span class="op">$</span>varComp</code></pre></div>
<pre><code>##   study_1   study_3   study_2 
##  98.20192 168.82045 155.70722</code></pre>
<p>The fitted values of the variance components are different for the different studies, indicating that the distributions of height in the three studies have different variance even after accounting for the other covariates.</p>
<p>We also recommend taking an inverse normal transform of the residuals and refitting the model. This is done separately for each group, and the transformed residuals are rescaled. See the full procedure in the<br />
<a href="https://github.com/UW-GAC/analysis_pipeline#association-testing">pipeline documenation</a>.</p>
</div>
<div id="single-variant-tests" class="section level2">
<h2><span class="header-section-number">4.2</span> Single-variant tests</h2>
<p>Now that we have a null model adjusting height for covariates, we can run an association test to look for genetic effects on height.</p>
<p>Single-variant tests are the same as in GWAS. We use the <code>assocTestSingle</code> function in GENESIS. First, we have to create a <code>SeqVarData</code> object including both the GDS file and the sample annotation containing phenotypes. We then create a <code>SeqVarBlockIterator</code> object to iterate over blocks of variants.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SeqVarTools)
gdsfile &lt;-<span class="st"> &quot;data/1KG_phase3_subset_chr1.gds&quot;</span>
gds &lt;-<span class="st"> </span><span class="kw">seqOpen</span>(gdsfile)
seqData &lt;-<span class="st"> </span><span class="kw">SeqVarData</span>(gds, <span class="dt">sampleData=</span>annot)
iterator &lt;-<span class="st"> </span><span class="kw">SeqVarBlockIterator</span>(seqData, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
assoc &lt;-<span class="st"> </span><span class="kw">assocTestSingle</span>(iterator, nullmod)</code></pre></div>
<pre><code>## # of selected samples: 1,126</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(assoc)</code></pre></div>
<pre><code>##   variant.id chr     pos allele.index n.obs         freq      Score
## 1          1   1  970546            1  1126 0.0039964476 -0.1191236
## 2          2   1  985900            1  1126 0.0492895204 -1.6707553
## 3          3   1 1025045            1  1126 0.0004440497 -0.2795838
## 4          4   1 1265550            1  1126 0.0008880995 -0.1105487
## 5          5   1 1472676            1  1126 0.0071047957  0.3630992
## 6          6   1 1735725            1  1126 0.0022202487 -0.1300405
##    Score.SE Score.Stat  Score.pval
## 1 0.2577712 -0.4621292 0.643988703
## 2 0.8841849 -1.8895995 0.058811539
## 3 0.1007173 -2.7759261 0.005504472
## 4 0.1085480 -1.0184319 0.308472754
## 5 0.3456555  1.0504657 0.293504072
## 6 0.1973175 -0.6590420 0.509868791</code></pre>
<p>We make a QQ plot to examine the results.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)
qqPlot &lt;-<span class="st"> </span><span class="cf">function</span>(pval) {
    pval &lt;-<span class="st"> </span>pval[<span class="op">!</span><span class="kw">is.na</span>(pval)]
    n &lt;-<span class="st"> </span><span class="kw">length</span>(pval)
    x &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span>n
    dat &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">obs=</span><span class="kw">sort</span>(pval),
                      <span class="dt">exp=</span>x<span class="op">/</span>n,
                      <span class="dt">upper=</span><span class="kw">qbeta</span>(<span class="fl">0.025</span>, x, <span class="kw">rev</span>(x)),
                      <span class="dt">lower=</span><span class="kw">qbeta</span>(<span class="fl">0.975</span>, x, <span class="kw">rev</span>(x)))
    
    <span class="kw">ggplot</span>(dat, <span class="kw">aes</span>(<span class="op">-</span><span class="kw">log10</span>(exp), <span class="op">-</span><span class="kw">log10</span>(obs))) <span class="op">+</span>
<span class="st">        </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(<span class="op">-</span><span class="kw">log10</span>(exp), <span class="op">-</span><span class="kw">log10</span>(upper)), <span class="dt">color=</span><span class="st">&quot;gray&quot;</span>) <span class="op">+</span>
<span class="st">        </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(<span class="op">-</span><span class="kw">log10</span>(exp), <span class="op">-</span><span class="kw">log10</span>(lower)), <span class="dt">color=</span><span class="st">&quot;gray&quot;</span>) <span class="op">+</span>
<span class="st">        </span><span class="kw">geom_point</span>() <span class="op">+</span>
<span class="st">        </span><span class="kw">geom_abline</span>(<span class="dt">intercept=</span><span class="dv">0</span>, <span class="dt">slope=</span><span class="dv">1</span>, <span class="dt">color=</span><span class="st">&quot;red&quot;</span>) <span class="op">+</span>
<span class="st">        </span><span class="kw">xlab</span>(<span class="kw">expression</span>(<span class="kw">paste</span>(<span class="op">-</span>log[<span class="dv">10</span>], <span class="st">&quot;(expected P)&quot;</span>))) <span class="op">+</span>
<span class="st">        </span><span class="kw">ylab</span>(<span class="kw">expression</span>(<span class="kw">paste</span>(<span class="op">-</span>log[<span class="dv">10</span>], <span class="st">&quot;(observed P)&quot;</span>))) <span class="op">+</span>
<span class="st">        </span><span class="kw">theme_bw</span>()
}    

<span class="kw">qqPlot</span>(assoc<span class="op">$</span>Score.pval)</code></pre></div>
<p><img src="_main_files/figure-html/assoc_single_qq-1.png" width="672" /></p>
</div>
<div id="exercises-1" class="section level2">
<h2><span class="header-section-number">4.3</span> Exercises</h2>
<ol style="list-style-type: decimal">
<li><p>Logistic regression: <code>fitNullModel</code> can use a binary phenotype as the outcome variable by specifying the argument <code>family=binomial</code>. Use the <code>status</code> column in the sample annotation to fit a null model for simulated case/control status, with <code>sex</code> and <code>Population</code> as covariates. Then run a single-variant test using this model.</p></li>
<li><p>Inverse normal transform: use the function <code>nullModelInvNorm</code> to perform an inverse normal transform on the <code>height</code> variable. For each study separately, compute a null model and do the inverse normal transform using just the values for that study. Compare these residuals with the initial residuals you obtained for that study by transforming all studies together.</p></li>
</ol>
</div>
<div id="sliding-window-tests" class="section level2">
<h2><span class="header-section-number">4.4</span> Sliding window tests</h2>
<p>For rare variants, we can do burden tests or SKAT using the GENESIS function <code>assocTestAggregate</code>. We restrict the test to variants with alternate allele frequency &lt; 0.1. (For real data, this threshold would be lower.) We use a flat weighting scheme. We define a sliding window across the genome using a <code>SeqVarWindowIterator</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqResetFilter</span>(seqData, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
iterator &lt;-<span class="st"> </span><span class="kw">SeqVarWindowIterator</span>(seqData, <span class="dt">windowSize=</span><span class="dv">5000</span>, <span class="dt">windowShift=</span><span class="dv">2000</span>, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
assoc &lt;-<span class="st"> </span><span class="kw">assocTestAggregate</span>(iterator, nullmod, <span class="dt">test=</span><span class="st">&quot;Burden&quot;</span>, <span class="dt">AF.max=</span><span class="fl">0.1</span>, <span class="dt">weight.beta=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</code></pre></div>
<pre><code>## # of selected samples: 1,126</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(assoc)</code></pre></div>
<pre><code>## [1] &quot;results&quot;     &quot;variantInfo&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(assoc<span class="op">$</span>results)</code></pre></div>
<pre><code>##   chr   start     end n.site n.alt n.sample.alt      Score  Score.SE
## 1   1  966001  971000      1     9            9 -0.1191236 0.2577712
## 2   1  982001  987000      1   111          107 -1.6707553 0.8841849
## 3   1 1022001 1027000      1     1            1 -0.2795838 0.1007173
## 4   1 1262001 1267000      1     2            2 -0.1105487 0.1085480
## 5   1 1468001 1473000      1    16           16  0.3630992 0.3456555
## 6   1 1732001 1737000      1     5            5 -0.1300405 0.1973175
##   Score.Stat  Score.pval
## 1 -0.4621292 0.643988703
## 2 -1.8895995 0.058811539
## 3 -2.7759261 0.005504472
## 4 -1.0184319 0.308472754
## 5  1.0504657 0.293504072
## 6 -0.6590420 0.509868791</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(assoc<span class="op">$</span>variantInfo)</code></pre></div>
<pre><code>## [[1]]
##   variant.id chr    pos allele.index n.obs        freq weight
## 1          1   1 970546            1  1126 0.003996448      1
## 
## [[2]]
##   variant.id chr    pos allele.index n.obs       freq weight
## 1          2   1 985900            1  1126 0.04928952      1
## 
## [[3]]
##   variant.id chr     pos allele.index n.obs         freq weight
## 1          3   1 1025045            1  1126 0.0004440497      1
## 
## [[4]]
##   variant.id chr     pos allele.index n.obs         freq weight
## 1          4   1 1265550            1  1126 0.0008880995      1
## 
## [[5]]
##   variant.id chr     pos allele.index n.obs        freq weight
## 1          5   1 1472676            1  1126 0.007104796      1
## 
## [[6]]
##   variant.id chr     pos allele.index n.obs        freq weight
## 1          6   1 1735725            1  1126 0.002220249      1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">qqPlot</span>(assoc<span class="op">$</span>results<span class="op">$</span>Score.pval)</code></pre></div>
<p><img src="_main_files/figure-html/assoc_window_burden-1.png" width="672" /></p>
<p>For SKAT, we use the Wu weights.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqResetFilter</span>(seqData, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
iterator &lt;-<span class="st"> </span><span class="kw">SeqVarWindowIterator</span>(seqData, <span class="dt">windowSize=</span><span class="dv">5000</span>, <span class="dt">windowShift=</span><span class="dv">2000</span>, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
assoc &lt;-<span class="st"> </span><span class="kw">assocTestAggregate</span>(iterator, nullmod, <span class="dt">test=</span><span class="st">&quot;SKAT&quot;</span>, <span class="dt">AF.max=</span><span class="fl">0.1</span>, <span class="dt">weight.beta=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">25</span>))</code></pre></div>
<pre><code>## # of selected samples: 1,126</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(assoc<span class="op">$</span>results)</code></pre></div>
<pre><code>##   chr   start     end n.site n.alt n.sample.alt        Q_0      pval_0
## 1   1  966001  971000      1     9            9   7.318094 0.643988703
## 2   1  982001  987000      1   111          107 154.178280 0.058811539
## 3   1 1022001 1027000      1     1            1  47.823916 0.005504472
## 4   1 1262001 1267000      1     2            2   7.319239 0.308472754
## 5   1 1468001 1473000      1    16           16  58.518662 0.293504072
## 6   1 1732001 1737000      1     5            5   9.499539 0.509868791
##   err_0
## 1     0
## 2     0
## 3     0
## 4     0
## 5     0
## 6     0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(assoc<span class="op">$</span>variantInfo)</code></pre></div>
<pre><code>## [[1]]
##   variant.id chr    pos allele.index n.obs        freq   weight
## 1          1   1 970546            1  1126 0.003996448 22.70917
## 
## [[2]]
##   variant.id chr    pos allele.index n.obs       freq   weight
## 1          2   1 985900            1  1126 0.04928952 7.431881
## 
## [[3]]
##   variant.id chr     pos allele.index n.obs         freq   weight
## 1          3   1 1025045            1  1126 0.0004440497 24.73493
## 
## [[4]]
##   variant.id chr     pos allele.index n.obs         freq   weight
## 1          4   1 1265550            1  1126 0.0008880995 24.47255
## 
## [[5]]
##   variant.id chr     pos allele.index n.obs        freq   weight
## 1          5   1 1472676            1  1126 0.007104796 21.06793
## 
## [[6]]
##   variant.id chr     pos allele.index n.obs        freq   weight
## 1          6   1 1735725            1  1126 0.002220249 23.70132</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">qqPlot</span>(assoc<span class="op">$</span>results<span class="op">$</span>pval_<span class="dv">0</span>)</code></pre></div>
<p><img src="_main_files/figure-html/assoc_window_skat-1.png" width="672" /></p>
</div>
<div id="exercise" class="section level2">
<h2><span class="header-section-number">4.5</span> Exercise</h2>
<ol start="3" style="list-style-type: decimal">
<li>Repeat the previous exercise on logistic regression, this time running a sliding-window test.</li>
</ol>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="gds-solutions.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="association-tests-solutions.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
