<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>3 GDS - Solutions | SISG Module 17: Computational Pipeline for WGS Data</title>
  <meta name="description" content="Course materials for SISG Module 17" />
  <meta name="generator" content="bookdown 0.11 and GitBook 2.6.7" />

  <meta property="og:title" content="3 GDS - Solutions | SISG Module 17: Computational Pipeline for WGS Data" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Course materials for SISG Module 17" />
  <meta name="github-repo" content="UW-GAC/SISG_2019" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3 GDS - Solutions | SISG Module 17: Computational Pipeline for WGS Data" />
  
  <meta name="twitter:description" content="Course materials for SISG Module 17" />
  



<meta name="date" content="2019-07-23" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="gds-format.html">
<link rel="next" href="phenotype-harmonization.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#schedule"><i class="fa fa-check"></i><b>1.1</b> Schedule</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#r-packages-used"><i class="fa fa-check"></i><b>1.2</b> R packages used</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#resources"><i class="fa fa-check"></i><b>1.3</b> Resources</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="gds-format.html"><a href="gds-format.html"><i class="fa fa-check"></i><b>2</b> GDS format</a><ul>
<li class="chapter" data-level="2.1" data-path="gds-format.html"><a href="gds-format.html#exploring-a-gds-file"><i class="fa fa-check"></i><b>2.1</b> Exploring a GDS file</a></li>
<li class="chapter" data-level="2.2" data-path="gds-format.html"><a href="gds-format.html#exercises"><i class="fa fa-check"></i><b>2.2</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="gds-solutions.html"><a href="gds-solutions.html"><i class="fa fa-check"></i><b>3</b> GDS - Solutions</a></li>
<li class="chapter" data-level="4" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html"><i class="fa fa-check"></i><b>4</b> Phenotype Harmonization</a><ul>
<li class="chapter" data-level="4.1" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#inspect-individual-study-data-in-r"><i class="fa fa-check"></i><b>4.1</b> Inspect individual study data in R</a></li>
<li class="chapter" data-level="4.2" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#compare-study-values"><i class="fa fa-check"></i><b>4.2</b> Compare study values</a></li>
<li class="chapter" data-level="4.3" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#using-null-models-to-compare-studies"><i class="fa fa-check"></i><b>4.3</b> Using null models to compare studies</a><ul>
<li class="chapter" data-level="4.3.1" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#create-an-annotated-data-frame"><i class="fa fa-check"></i><b>4.3.1</b> Create an Annotated Data Frame</a></li>
<li class="chapter" data-level="4.3.2" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#fit-a-null-model-without-study"><i class="fa fa-check"></i><b>4.3.2</b> Fit a null model without study</a></li>
<li class="chapter" data-level="4.3.3" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#fit-a-model-with-study"><i class="fa fa-check"></i><b>4.3.3</b> Fit a model with study</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#final-considerations"><i class="fa fa-check"></i><b>4.4</b> Final considerations</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="association-tests.html"><a href="association-tests.html"><i class="fa fa-check"></i><b>5</b> Association tests</a><ul>
<li class="chapter" data-level="5.1" data-path="association-tests.html"><a href="association-tests.html#null-model"><i class="fa fa-check"></i><b>5.1</b> Null model</a></li>
<li class="chapter" data-level="5.2" data-path="association-tests.html"><a href="association-tests.html#single-variant-tests"><i class="fa fa-check"></i><b>5.2</b> Single-variant tests</a></li>
<li class="chapter" data-level="5.3" data-path="association-tests.html"><a href="association-tests.html#exercises-1"><i class="fa fa-check"></i><b>5.3</b> Exercises</a></li>
<li class="chapter" data-level="5.4" data-path="association-tests.html"><a href="association-tests.html#sliding-window-tests"><i class="fa fa-check"></i><b>5.4</b> Sliding window tests</a></li>
<li class="chapter" data-level="5.5" data-path="association-tests.html"><a href="association-tests.html#exercise"><i class="fa fa-check"></i><b>5.5</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="association-tests-solutions.html"><a href="association-tests-solutions.html"><i class="fa fa-check"></i><b>6</b> Association tests - Solutions</a></li>
<li class="chapter" data-level="7" data-path="computing-a-grm.html"><a href="computing-a-grm.html"><i class="fa fa-check"></i><b>7</b> Computing a GRM</a></li>
<li class="chapter" data-level="8" data-path="pc-relate.html"><a href="pc-relate.html"><i class="fa fa-check"></i><b>8</b> PC-Relate</a><ul>
<li class="chapter" data-level="8.1" data-path="pc-relate.html"><a href="pc-relate.html#king"><i class="fa fa-check"></i><b>8.1</b> KING</a></li>
<li class="chapter" data-level="8.2" data-path="pc-relate.html"><a href="pc-relate.html#pc-air"><i class="fa fa-check"></i><b>8.2</b> PC-AiR</a></li>
<li class="chapter" data-level="8.3" data-path="pc-relate.html"><a href="pc-relate.html#pc-relate-1"><i class="fa fa-check"></i><b>8.3</b> PC-Relate</a></li>
<li class="chapter" data-level="8.4" data-path="pc-relate.html"><a href="pc-relate.html#comparison-with-pedigree"><i class="fa fa-check"></i><b>8.4</b> Comparison with pedigree</a></li>
<li class="chapter" data-level="8.5" data-path="pc-relate.html"><a href="pc-relate.html#exercise-1"><i class="fa fa-check"></i><b>8.5</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="mixed-models.html"><a href="mixed-models.html"><i class="fa fa-check"></i><b>9</b> Mixed models</a><ul>
<li class="chapter" data-level="9.1" data-path="mixed-models.html"><a href="mixed-models.html#null-model-1"><i class="fa fa-check"></i><b>9.1</b> Null model</a></li>
<li class="chapter" data-level="9.2" data-path="mixed-models.html"><a href="mixed-models.html#single-variant-tests-1"><i class="fa fa-check"></i><b>9.2</b> Single-variant tests</a></li>
<li class="chapter" data-level="9.3" data-path="mixed-models.html"><a href="mixed-models.html#exercise-2"><i class="fa fa-check"></i><b>9.3</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="mixed-models-solutions.html"><a href="mixed-models-solutions.html"><i class="fa fa-check"></i><b>10</b> Mixed models - Solutions</a></li>
<li class="chapter" data-level="11" data-path="variant-annotation.html"><a href="variant-annotation.html"><i class="fa fa-check"></i><b>11</b> Variant annotation</a><ul>
<li class="chapter" data-level="11.1" data-path="variant-annotation.html"><a href="variant-annotation.html#working-with-variant-annotation"><i class="fa fa-check"></i><b>11.1</b> Working with variant annotation</a><ul>
<li class="chapter" data-level="11.1.1" data-path="variant-annotation.html"><a href="variant-annotation.html#wgsaparsr-configuration"><i class="fa fa-check"></i><b>11.1.1</b> WGSAParsr Configuration</a></li>
<li class="chapter" data-level="11.1.2" data-path="variant-annotation.html"><a href="variant-annotation.html#parsing-with-wgsaparsr"><i class="fa fa-check"></i><b>11.1.2</b> Parsing with WGSAParsr</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="variant-annotation.html"><a href="variant-annotation.html#aggregating-and-filtering-variants-using-annotation"><i class="fa fa-check"></i><b>11.2</b> Aggregating and filtering variants using annotation</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="aggregate-tests.html"><a href="aggregate-tests.html"><i class="fa fa-check"></i><b>12</b> Aggregate tests</a><ul>
<li class="chapter" data-level="12.1" data-path="aggregate-tests.html"><a href="aggregate-tests.html#aggregate-unit-for-association-testing-exercise"><i class="fa fa-check"></i><b>12.1</b> Aggregate unit for association testing exercise</a></li>
<li class="chapter" data-level="12.2" data-path="aggregate-tests.html"><a href="aggregate-tests.html#association-testing-with-aggregate-units"><i class="fa fa-check"></i><b>12.2</b> Association testing with aggregate units</a></li>
<li class="chapter" data-level="12.3" data-path="aggregate-tests.html"><a href="aggregate-tests.html#exercise-3"><i class="fa fa-check"></i><b>12.3</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="annotation-solutions.html"><a href="annotation-solutions.html"><i class="fa fa-check"></i><b>13</b> Annotation - Solutions</a></li>
<li class="chapter" data-level="14" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html"><i class="fa fa-check"></i><b>14</b> Analysis Pipeline</a><ul>
<li class="chapter" data-level="14.1" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#running-on-a-local-cluster"><i class="fa fa-check"></i><b>14.1</b> Running on a local cluster</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="analysis-commons.html"><a href="analysis-commons.html"><i class="fa fa-check"></i><b>15</b> Analysis Commons</a><ul>
<li class="chapter" data-level="15.1" data-path="analysis-commons.html"><a href="analysis-commons.html#outline"><i class="fa fa-check"></i><b>15.1</b> Outline</a></li>
<li class="chapter" data-level="15.2" data-path="analysis-commons.html"><a href="analysis-commons.html#web-interface-and-running-an-analysis-application"><i class="fa fa-check"></i><b>15.2</b> Web Interface and Running an Analysis Application</a><ul>
<li class="chapter" data-level="15.2.1" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-1-run-a-single-variant-analysis."><i class="fa fa-check"></i><b>15.2.1</b> Exercise 1) Run a single variant analysis.</a></li>
<li class="chapter" data-level="15.2.2" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-2-run-skat-test-grouping-variants-into-gene-transcript-regions-and-limit-the-variants-to-those-with-a-cadd-phred-score-2-and-maf-5."><i class="fa fa-check"></i><b>15.2.2</b> Exercise 2) Run SKAT test grouping variants into gene transcript regions and limit the variants to those with a CADD phred score &gt; 2 and MAF &lt;= 5%.</a></li>
</ul></li>
<li class="chapter" data-level="15.3" data-path="analysis-commons.html"><a href="analysis-commons.html#command-line-interface"><i class="fa fa-check"></i><b>15.3</b> Command line interface</a><ul>
<li class="chapter" data-level="15.3.1" data-path="analysis-commons.html"><a href="analysis-commons.html#log-in-to-awi"><i class="fa fa-check"></i><b>15.3.1</b> Log in to AWI</a></li>
<li class="chapter" data-level="15.3.2" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-3-navigate-directories-make-output-directory-examine-files"><i class="fa fa-check"></i><b>15.3.2</b> Exercise 3) Navigate directories, make output directory, examine files</a></li>
<li class="chapter" data-level="15.3.3" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-4-run-single-variant-analysis-from-command-line-using-bash-script"><i class="fa fa-check"></i><b>15.3.3</b> Exercise 4) Run single variant analysis from command line using bash script</a></li>
</ul></li>
<li class="chapter" data-level="15.4" data-path="analysis-commons.html"><a href="analysis-commons.html#writing-your-own-apps"><i class="fa fa-check"></i><b>15.4</b> Writing your own Apps</a><ul>
<li class="chapter" data-level="15.4.1" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-5-write-an-app-that-creates-phenotype-residuals-and-performs-an-inverse-normal-transform"><i class="fa fa-check"></i><b>15.4.1</b> Exercise 5) Write an App that creates phenotype residuals and performs an inverse normal transform</a></li>
<li class="chapter" data-level="15.4.2" data-path="analysis-commons.html"><a href="analysis-commons.html#optional-exercise-6-make-qq-plot"><i class="fa fa-check"></i><b>15.4.2</b> Optional Exercise 6) Make QQ plot</a></li>
<li class="chapter" data-level="15.4.3" data-path="analysis-commons.html"><a href="analysis-commons.html#optional-exercise-8-create-a-regional-association-plot-using-ld-extracted-from-your-data-set"><i class="fa fa-check"></i><b>15.4.3</b> Optional Exercise 8) Create a regional association plot using LD extracted from your data set</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">SISG Module 17: Computational Pipeline for WGS Data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="gds---solutions" class="section level1">
<h1><span class="header-section-number">3</span> GDS - Solutions</h1>
<ol style="list-style-type: decimal">
<li>Set a filter selecting only multi-allelic variants. Inspect their genotypes using the different methods you learned above. Use the <code>alleleDosage</code> method to find dosage for the second (and third, etc.) alternate allele.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqSetFilter</span>(gds, <span class="dt">variant.sel=</span>multi.allelic)</code></pre></div>
<pre><code>## # of selected variants: 21</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">geno &lt;-<span class="st"> </span><span class="kw">seqGetData</span>(gds, <span class="st">&quot;genotype&quot;</span>)
<span class="kw">dim</span>(geno)</code></pre></div>
<pre><code>## [1]    2 1126   21</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">geno[,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,]</code></pre></div>
<pre><code>## , , 1
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    1    1    0    1    1
##   [2,]    0    1    1    1    1
## 
## , , 2
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    0
##   [2,]    0    0    0    0    0
## 
## , , 3
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    0
##   [2,]    0    0    0    0    0
## 
## , , 4
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    1    0    0    1    0
##   [2,]    0    0    0    0    1
## 
## , , 5
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    0
##   [2,]    0    0    0    0    0
## 
## , , 6
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    0
##   [2,]    0    0    0    0    0
## 
## , , 7
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    0
##   [2,]    0    0    0    0    0
## 
## , , 8
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    0
##   [2,]    0    0    0    0    0
## 
## , , 9
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    2
##   [2,]    0    0    0    0    0
## 
## , , 10
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    1    2    0    0    1
##   [2,]    0    2    0    2    1
## 
## , , 11
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    1    0    0
##   [2,]    1    1    0    1    0
## 
## , , 12
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    3    1    1    1    3
##   [2,]    3    3    3    1    0
## 
## , , 13
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    2    0
##   [2,]    0    0    0    0    0
## 
## , , 14
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    0
##   [2,]    0    0    0    0    0
## 
## , , 15
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    2    0    0    0    2
##   [2,]    0    0    0    2    0
## 
## , , 16
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    2    2    2    2    2
##   [2,]    2    2    2    2    2
## 
## , , 17
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    2    2    2    2    2
##   [2,]    2    2    2    2    2
## 
## , , 18
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    0
##   [2,]    0    0    0    0    0
## 
## , , 19
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    0
##   [2,]    0    0    0    0    0
## 
## , , 20
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    1
##   [2,]    0    0    0    0    0
## 
## , , 21
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    0
##   [2,]    0    0    0    2    0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">geno &lt;-<span class="st"> </span><span class="kw">getGenotype</span>(gds)
<span class="kw">dim</span>(geno)</code></pre></div>
<pre><code>## [1] 1126   21</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(geno)</code></pre></div>
<pre><code>##          variant
## sample    30    69    73    161   162   195   243   253   407   431  
##   HG00096 &quot;1|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;1|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;1|0&quot;
##   HG00097 &quot;1|1&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;2|2&quot;
##   HG00099 &quot;0|1&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot;
##   HG00100 &quot;1|1&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;1|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|2&quot;
##   HG00101 &quot;1|1&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|1&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;2|0&quot; &quot;1|1&quot;
##   HG00102 &quot;0|1&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;1|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;2|2&quot;
##          variant
## sample    434   610   627   645   689   756   765   814   988   1014 
##   HG00096 &quot;0|1&quot; &quot;3|3&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;2|0&quot; &quot;2|2&quot; &quot;2|2&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot;
##   HG00097 &quot;0|1&quot; &quot;1|3&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;2|2&quot; &quot;2|2&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot;
##   HG00099 &quot;1|0&quot; &quot;1|3&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;2|2&quot; &quot;2|2&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot;
##   HG00100 &quot;0|1&quot; &quot;1|1&quot; &quot;2|0&quot; &quot;0|0&quot; &quot;0|2&quot; &quot;2|2&quot; &quot;2|2&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot;
##   HG00101 &quot;0|0&quot; &quot;3|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;2|0&quot; &quot;2|2&quot; &quot;2|2&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;1|0&quot;
##   HG00102 &quot;0|1&quot; &quot;3|3&quot; &quot;0|2&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;2|2&quot; &quot;1|2&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|1&quot;
##          variant
## sample    1056 
##   HG00096 &quot;0|0&quot;
##   HG00097 &quot;0|0&quot;
##   HG00099 &quot;0|0&quot;
##   HG00100 &quot;0|2&quot;
##   HG00101 &quot;0|0&quot;
##   HG00102 &quot;0|2&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">geno &lt;-<span class="st"> </span><span class="kw">getGenotypeAlleles</span>(gds)
<span class="kw">head</span>(geno)</code></pre></div>
<pre><code>##          variant
## sample    30       69    73    161   162   195   243   253   407  
##   HG00096 &quot;GT|GTT&quot; &quot;A|A&quot; &quot;G|G&quot; &quot;A|G&quot; &quot;A|A&quot; &quot;A|A&quot; &quot;C|C&quot; &quot;G|G&quot; &quot;T|T&quot;
##   HG00097 &quot;GT|GT&quot;  &quot;A|A&quot; &quot;G|G&quot; &quot;G|G&quot; &quot;A|A&quot; &quot;A|A&quot; &quot;C|C&quot; &quot;G|G&quot; &quot;T|T&quot;
##   HG00099 &quot;GTT|GT&quot; &quot;A|A&quot; &quot;G|G&quot; &quot;G|G&quot; &quot;A|A&quot; &quot;A|A&quot; &quot;C|C&quot; &quot;G|G&quot; &quot;T|T&quot;
##   HG00100 &quot;GT|GT&quot;  &quot;A|A&quot; &quot;G|G&quot; &quot;A|G&quot; &quot;A|A&quot; &quot;A|A&quot; &quot;C|C&quot; &quot;G|G&quot; &quot;T|T&quot;
##   HG00101 &quot;GT|GT&quot;  &quot;A|A&quot; &quot;G|G&quot; &quot;G|A&quot; &quot;A|A&quot; &quot;A|A&quot; &quot;C|C&quot; &quot;G|G&quot; &quot;C|T&quot;
##   HG00102 &quot;GTT|GT&quot; &quot;A|A&quot; &quot;G|G&quot; &quot;A|G&quot; &quot;A|A&quot; &quot;A|A&quot; &quot;C|C&quot; &quot;G|G&quot; &quot;T|T&quot;
##          variant
## sample    431       434      610       627   645   689   756              
##   HG00096 &quot;TAA|TA&quot;  &quot;G|GTTA&quot; &quot;G|G&quot;     &quot;G|G&quot; &quot;G|G&quot; &quot;C|G&quot; &quot;CGAGCAT|CGAGCAT&quot;
##   HG00097 &quot;T|T&quot;     &quot;G|GTTA&quot; &quot;GCC|G&quot;   &quot;G|G&quot; &quot;G|G&quot; &quot;G|G&quot; &quot;CGAGCAT|CGAGCAT&quot;
##   HG00099 &quot;TA|TA&quot;   &quot;GTTA|G&quot; &quot;GCC|G&quot;   &quot;G|G&quot; &quot;G|G&quot; &quot;G|G&quot; &quot;CGAGCAT|CGAGCAT&quot;
##   HG00100 &quot;TA|T&quot;    &quot;G|GTTA&quot; &quot;GCC|GCC&quot; &quot;C|G&quot; &quot;G|G&quot; &quot;G|C&quot; &quot;CGAGCAT|CGAGCAT&quot;
##   HG00101 &quot;TAA|TAA&quot; &quot;G|G&quot;    &quot;G|GC&quot;    &quot;G|G&quot; &quot;G|G&quot; &quot;C|G&quot; &quot;CGAGCAT|CGAGCAT&quot;
##   HG00102 &quot;T|T&quot;     &quot;G|GTTA&quot; &quot;G|G&quot;     &quot;G|C&quot; &quot;G|G&quot; &quot;G|G&quot; &quot;CGAGCAT|CGAGCAT&quot;
##          variant
## sample    765         814     988           1014      1056 
##   HG00096 &quot;C|C&quot;       &quot;CT|CT&quot; &quot;CGTGA|CGTGA&quot; &quot;C|C&quot;     &quot;A|A&quot;
##   HG00097 &quot;C|C&quot;       &quot;CT|CT&quot; &quot;CGTGA|CGTGA&quot; &quot;C|C&quot;     &quot;A|A&quot;
##   HG00099 &quot;C|C&quot;       &quot;CT|CT&quot; &quot;CGTGA|CGTGA&quot; &quot;C|C&quot;     &quot;A|A&quot;
##   HG00100 &quot;C|C&quot;       &quot;CT|CT&quot; &quot;CGTGA|CGTGA&quot; &quot;C|C&quot;     &quot;A|G&quot;
##   HG00101 &quot;C|C&quot;       &quot;CT|CT&quot; &quot;CGTGA|CGTGA&quot; &quot;CCATT|C&quot; &quot;A|A&quot;
##   HG00102 &quot;CATTATT|C&quot; &quot;CT|CT&quot; &quot;CGTGA|CGTGA&quot; &quot;C|CCATT&quot; &quot;A|G&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dos &lt;-<span class="st"> </span><span class="kw">refDosage</span>(gds)
<span class="kw">head</span>(dos)</code></pre></div>
<pre><code>##          variant
## sample    30 69 73 161 162 195 243 253 407 431 434 610 627 645 689 756 765
##   HG00096  1  2  2   1   2   2   2   2   2   1   1   0   2   2   1   0   0
##   HG00097  0  2  2   2   2   2   2   2   2   0   1   0   2   2   2   0   0
##   HG00099  1  2  2   2   2   2   2   2   2   2   1   0   2   2   2   0   0
##   HG00100  0  2  2   1   2   2   2   2   2   1   1   0   1   2   1   0   0
##   HG00101  0  2  2   1   2   2   2   2   1   0   2   1   2   2   1   0   0
##   HG00102  1  2  2   1   2   2   2   2   2   0   1   0   1   2   2   0   0
##          variant
## sample    814 988 1014 1056
##   HG00096   2   2    2    2
##   HG00097   2   2    2    2
##   HG00099   2   2    2    2
##   HG00100   2   2    2    1
##   HG00101   2   2    1    2
##   HG00102   2   2    1    1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dos &lt;-<span class="st"> </span><span class="kw">altDosage</span>(gds)
<span class="kw">head</span>(dos)</code></pre></div>
<pre><code>##          variant
## sample    30 69 73 161 162 195 243 253 407 431 434 610 627 645 689 756 765
##   HG00096  1  0  0   1   0   0   0   0   0   1   1   2   0   0   1   2   2
##   HG00097  2  0  0   0   0   0   0   0   0   2   1   2   0   0   0   2   2
##   HG00099  1  0  0   0   0   0   0   0   0   0   1   2   0   0   0   2   2
##   HG00100  2  0  0   1   0   0   0   0   0   1   1   2   1   0   1   2   2
##   HG00101  2  0  0   1   0   0   0   0   1   2   0   1   0   0   1   2   2
##   HG00102  1  0  0   1   0   0   0   0   0   2   1   2   1   0   0   2   2
##          variant
## sample    814 988 1014 1056
##   HG00096   0   0    0    0
##   HG00097   0   0    0    0
##   HG00099   0   0    0    0
##   HG00100   0   0    0    1
##   HG00101   0   0    1    0
##   HG00102   0   0    1    1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dos &lt;-<span class="st"> </span><span class="kw">alleleDosage</span>(gds, <span class="dt">n=</span><span class="dv">2</span>)
<span class="kw">head</span>(dos)</code></pre></div>
<pre><code>##          variant
## sample    30 69 73 161 162 195 243 253 407 431 434 610 627 645 689 756 765
##   HG00096  0  0  0   0   0   0   0   0   0   0   0   0   0   0   1   2   2
##   HG00097  0  0  0   0   0   0   0   0   0   2   0   0   0   0   0   2   2
##   HG00099  0  0  0   0   0   0   0   0   0   0   0   0   0   0   0   2   2
##   HG00100  0  0  0   0   0   0   0   0   0   1   0   0   1   0   1   2   2
##   HG00101  0  0  0   0   0   0   0   0   1   0   0   0   0   0   1   2   2
##   HG00102  0  0  0   0   0   0   0   0   0   2   0   0   1   0   0   2   1
##          variant
## sample    814 988 1014 1056
##   HG00096   0   0    0    0
##   HG00097   0   0    0    0
##   HG00099   0   0    0    0
##   HG00100   0   0    0    1
##   HG00101   0   0    0    0
##   HG00102   0   0    0    1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dos &lt;-<span class="st"> </span><span class="kw">alleleDosage</span>(gds, <span class="dt">n=</span><span class="dv">3</span>)
<span class="kw">head</span>(dos)</code></pre></div>
<pre><code>##          variant
## sample    30 69 73 161 162 195 243 253 407 431 434 610 627 645 689 756 765
##   HG00096  0  0  0   0   0   0   0   0   0   0   0   2   0   0   0   0   0
##   HG00097  0  0  0   0   0   0   0   0   0   0   0   1   0   0   0   0   0
##   HG00099  0  0  0   0   0   0   0   0   0   0   0   1   0   0   0   0   0
##   HG00100  0  0  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
##   HG00101  0  0  0   0   0   0   0   0   0   0   0   1   0   0   0   0   0
##   HG00102  0  0  0   0   0   0   0   0   0   0   0   2   0   0   0   0   0
##          variant
## sample    814 988 1014 1056
##   HG00096   0   0    0    0
##   HG00097   0   0    0    0
##   HG00099   0   0    0    0
##   HG00100   0   0    0    0
##   HG00101   0   0    0    0
##   HG00102   0   0    0    0</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Use the <code>hwe</code> function in SeqVarTools to run a Hardy-Weinberg Equilibrium test on each variant. Identify a variant with low p-value and inspect its genotypes. (Note that the HWE test is only valid for biallelic variants, and will return <code>NA</code> for multiallelic variants.)</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqResetFilter</span>(gds)</code></pre></div>
<pre><code>## # of selected samples: 1,126
## # of selected variants: 1,120</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hwe.res &lt;-<span class="st"> </span><span class="kw">hwe</span>(gds)
lowp &lt;-<span class="st"> </span><span class="op">!</span><span class="kw">is.na</span>(hwe.res<span class="op">$</span>p) <span class="op">&amp;</span><span class="st"> </span>hwe.res<span class="op">$</span>p <span class="op">&lt;</span><span class="st"> </span><span class="fl">1e-4</span>
<span class="kw">head</span>(hwe.res[lowp,])</code></pre></div>
<pre><code>##     variant.id nAA nAa naa     afreq            p         f
## 75          75 702 336  88 0.7726465 1.070663e-06 0.1506466
## 92          92 632 381 113 0.7304618 3.558878e-06 0.1407120
## 98          98 672 335 119 0.7455595 2.369695e-12 0.2158342
## 105        105  93 272 761 0.2033748 7.851777e-16 0.2544970
## 114        114 299 482 345 0.4795737 1.745346e-06 0.1424409
## 150        150 471 447 208 0.6167851 8.020208e-08 0.1602251</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqSetFilter</span>(gds, <span class="dt">variant.id=</span><span class="dv">75</span>)</code></pre></div>
<pre><code>## # of selected variants: 1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(<span class="kw">getGenotype</span>(gds))</code></pre></div>
<pre><code>## 
## 0|0 0|1 1|0 1|1 
## 702 165 171  88</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(<span class="kw">refDosage</span>(gds))</code></pre></div>
<pre><code>## 
##   0   1   2 
##  88 336 702</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqClose</span>(gds)</code></pre></div>

</div>
            </section>

          </div>
        </div>
      </div>
<a href="gds-format.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="phenotype-harmonization.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
