<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>2 GDS format | WISG Module 4: Computational Pipeline for WGS Data</title>
  <meta name="description" content="Course materials for WISG Module 4">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="2 GDS format | WISG Module 4: Computational Pipeline for WGS Data />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Course materials for WISG Module 4" />
  <meta name="github-repo" content="UW-GAC/WISG_2019" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="2 GDS format | WISG Module 4: Computational Pipeline for WGS Data />
  
  <meta name="twitter:description" content="Course materials for WISG Module 4" />
  



<meta name="date" content="2018-12-18">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="index.html">
<link rel="next" href="association-tests.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#software-requirements"><i class="fa fa-check"></i><b>1.1</b> Software requirements</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#resources"><i class="fa fa-check"></i><b>1.2</b> Resources</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="gds-format.html"><a href="gds-format.html"><i class="fa fa-check"></i><b>2</b> GDS format</a><ul>
<li class="chapter" data-level="2.1" data-path="gds-format.html"><a href="gds-format.html#exploring-a-gds-file"><i class="fa fa-check"></i><b>2.1</b> Exploring a GDS file</a></li>
<li class="chapter" data-level="2.2" data-path="gds-format.html"><a href="gds-format.html#exercises"><i class="fa fa-check"></i><b>2.2</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="association-tests.html"><a href="association-tests.html"><i class="fa fa-check"></i><b>3</b> Association tests</a><ul>
<li class="chapter" data-level="3.1" data-path="association-tests.html"><a href="association-tests.html#null-model"><i class="fa fa-check"></i><b>3.1</b> Null model</a></li>
<li class="chapter" data-level="3.2" data-path="association-tests.html"><a href="association-tests.html#single-variant-tests"><i class="fa fa-check"></i><b>3.2</b> Single-variant tests</a></li>
<li class="chapter" data-level="3.3" data-path="association-tests.html"><a href="association-tests.html#exercises-1"><i class="fa fa-check"></i><b>3.3</b> Exercises</a></li>
<li class="chapter" data-level="3.4" data-path="association-tests.html"><a href="association-tests.html#sliding-window-tests"><i class="fa fa-check"></i><b>3.4</b> Sliding window tests</a><ul>
<li class="chapter" data-level="3.4.1" data-path="association-tests.html"><a href="association-tests.html#exercise"><i class="fa fa-check"></i><b>3.4.1</b> Exercise</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="computing-a-grm.html"><a href="computing-a-grm.html"><i class="fa fa-check"></i><b>4</b> Computing a GRM</a></li>
<li class="chapter" data-level="5" data-path="pc-relate.html"><a href="pc-relate.html"><i class="fa fa-check"></i><b>5</b> PC-Relate</a><ul>
<li class="chapter" data-level="5.1" data-path="pc-relate.html"><a href="pc-relate.html#king"><i class="fa fa-check"></i><b>5.1</b> KING</a></li>
<li class="chapter" data-level="5.2" data-path="pc-relate.html"><a href="pc-relate.html#pc-air"><i class="fa fa-check"></i><b>5.2</b> PC-AiR</a></li>
<li class="chapter" data-level="5.3" data-path="pc-relate.html"><a href="pc-relate.html#pc-relate-1"><i class="fa fa-check"></i><b>5.3</b> PC-Relate</a></li>
<li class="chapter" data-level="5.4" data-path="pc-relate.html"><a href="pc-relate.html#comparison-with-pedigree"><i class="fa fa-check"></i><b>5.4</b> Comparison with pedigree</a></li>
<li class="chapter" data-level="5.5" data-path="pc-relate.html"><a href="pc-relate.html#exercise-1"><i class="fa fa-check"></i><b>5.5</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="mixed-models.html"><a href="mixed-models.html"><i class="fa fa-check"></i><b>6</b> Mixed models</a><ul>
<li class="chapter" data-level="6.1" data-path="mixed-models.html"><a href="mixed-models.html#null-model-1"><i class="fa fa-check"></i><b>6.1</b> Null model</a></li>
<li class="chapter" data-level="6.2" data-path="mixed-models.html"><a href="mixed-models.html#single-variant-tests-1"><i class="fa fa-check"></i><b>6.2</b> Single-variant tests</a></li>
<li class="chapter" data-level="6.3" data-path="mixed-models.html"><a href="mixed-models.html#exercise-2"><i class="fa fa-check"></i><b>6.3</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="variant-annotation.html"><a href="variant-annotation.html"><i class="fa fa-check"></i><b>7</b> Variant annotation</a><ul>
<li class="chapter" data-level="7.1" data-path="variant-annotation.html"><a href="variant-annotation.html#using-bioconductor-annotation-resources"><i class="fa fa-check"></i><b>7.1</b> Using Bioconductor annotation resources</a></li>
<li class="chapter" data-level="7.2" data-path="variant-annotation.html"><a href="variant-annotation.html#aggregating-and-filtering-variants-using-annotation"><i class="fa fa-check"></i><b>7.2</b> Aggregating and filtering variants using annotation</a></li>
<li class="chapter" data-level="7.3" data-path="variant-annotation.html"><a href="variant-annotation.html#aggregate-unit-for-association-testing-exercise"><i class="fa fa-check"></i><b>7.3</b> Aggregate unit for association testing exercise</a></li>
<li class="chapter" data-level="7.4" data-path="variant-annotation.html"><a href="variant-annotation.html#association-testing-with-aggregate-units"><i class="fa fa-check"></i><b>7.4</b> Association testing with aggregate units</a></li>
<li class="chapter" data-level="7.5" data-path="variant-annotation.html"><a href="variant-annotation.html#exercise-3"><i class="fa fa-check"></i><b>7.5</b> Exercise</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">WISG Module 4: Computational Pipeline for WGS Data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="gds-format" class="section level1">
<h1><span class="header-section-number">2</span> GDS format</h1>
<p>GDS is Genomic Data Structure, a storage format that can efficiently store genomic data and provide fast random access to subsets of the data. For more information on GDS for sequence data, read the <a href="https://github.com/zhengxwen/SeqArray/blob/master/vignettes/SeqArrayTutorial.Rmd">SeqArray package vignette</a>.</p>
<div id="exploring-a-gds-file" class="section level2">
<h2><span class="header-section-number">2.1</span> Exploring a GDS file</h2>
<p>To use the R packages developed at the University of Washington for sequence data, we first need to convert a VCF file to GDS. (If the file is BCF, use <a href="bcftools">https://samtools.github.io/bcftools/bcftools.html</a> to convert to VCF.)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SeqArray)
vcffile &lt;-<span class="st"> &quot;data/1KG_phase3_subset_chr1.vcf.gz&quot;</span>
gdsfile &lt;-<span class="st"> &quot;1KG_phase3_subset_chr1.gds&quot;</span>
<span class="kw">seqVCF2GDS</span>(vcffile, gdsfile, <span class="dt">fmt.import=</span><span class="st">&quot;GT&quot;</span>, <span class="dt">storage.option=</span><span class="st">&quot;LZMA_RA&quot;</span>, <span class="dt">verbose=</span><span class="ot">FALSE</span>)</code></pre></div>
<p>We can interact with the GDS file using the SeqArray package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gds &lt;-<span class="st"> </span><span class="kw">seqOpen</span>(gdsfile)
gds</code></pre></div>
<pre><code>## Object of class &quot;SeqVarGDSClass&quot;
## File: /Users/stephanie/WISG_2019/WISG_2019/1KG_phase3_subset_chr1.gds (70.6K)
## +    [  ] *
## |--+ description   [  ] *
## |--+ sample.id   { Str8 1126 LZMA_ra(9.66%), 877B } *
## |--+ variant.id   { Int32 1120 LZMA_ra(17.5%), 793B } *
## |--+ position   { Int32 1120 LZMA_ra(78.5%), 3.4K } *
## |--+ chromosome   { Str8 1120 LZMA_ra(4.55%), 109B } *
## |--+ allele   { Str8 1120 LZMA_ra(26.0%), 1.2K } *
## |--+ genotype   [  ] *
## |  |--+ data   { Bit2 2x1126x1121 LZMA_ra(8.34%), 51.4K } *
## |  |--+ extra.index   { Int32 3x0 LZMA_ra, 18B } *
## |  \--+ extra   { Int16 0 LZMA_ra, 18B }
## |--+ phase   [  ]
## |  |--+ data   { Bit1 1126x1120 LZMA_ra(0.11%), 177B } *
## |  |--+ extra.index   { Int32 3x0 LZMA_ra, 18B } *
## |  \--+ extra   { Bit1 0 LZMA_ra, 18B }
## |--+ annotation   [  ]
## |  |--+ id   { Str8 1120 LZMA_ra(40.4%), 3.6K } *
## |  |--+ qual   { Float32 1120 LZMA_ra(2.46%), 117B } *
## |  |--+ filter   { Int32,factor 1120 LZMA_ra(2.46%), 117B } *
## |  |--+ info   [  ]
## |  \--+ format   [  ]
## \--+ sample.annotation   [  ]</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># the unique sample identifier comes from the VCF header</span>
sample.id &lt;-<span class="st"> </span><span class="kw">seqGetData</span>(gds, <span class="st">&quot;sample.id&quot;</span>)
<span class="kw">length</span>(sample.id)</code></pre></div>
<pre><code>## [1] 1126</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(sample.id)</code></pre></div>
<pre><code>## [1] &quot;HG00096&quot; &quot;HG00097&quot; &quot;HG00099&quot; &quot;HG00100&quot; &quot;HG00101&quot; &quot;HG00102&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># a unique integer ID is assigned to each variant</span>
variant.id &lt;-<span class="st"> </span><span class="kw">seqGetData</span>(gds, <span class="st">&quot;variant.id&quot;</span>)
<span class="kw">length</span>(variant.id)</code></pre></div>
<pre><code>## [1] 1120</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(variant.id)</code></pre></div>
<pre><code>## [1] 1 2 3 4 5 6</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># reference allele frequency of each variant</span>
afreq &lt;-<span class="st"> </span><span class="kw">seqAlleleFreq</span>(gds)
<span class="kw">hist</span>(afreq, <span class="dt">breaks=</span><span class="dv">50</span>)</code></pre></div>
<p><img src="_main_files/figure-html/seqarray-1.png" width="672" /></p>
<p>We can define a filter on the <code>gds</code> object. After using the <code>seqSetFilter</code> command, all subsequent reads from the <code>gds</code> object are restricted to the selected subset of data, until a new filter is defined or <code>seqResetFilter</code> is called.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqSetFilter</span>(gds, <span class="dt">variant.id=</span><span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">sample.id=</span>sample.id[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>])</code></pre></div>
<pre><code>## # of selected samples: 5
## # of selected variants: 10</code></pre>
<p>Genotype data is stored in a 3-dimensional array, where the first dimension is always 2 for diploid genotypes. The second and third dimensions are samples and variants, respectively. The values of the array denote alleles: <code>0</code> is the reference allele and <code>1</code> is the alternate allele. For multiallelic variants, other alternate alleles are represented as integers <code>&gt; 1</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">geno &lt;-<span class="st"> </span><span class="kw">seqGetData</span>(gds, <span class="st">&quot;genotype&quot;</span>)
<span class="kw">dim</span>(geno)</code></pre></div>
<pre><code>## [1]  2  5 10</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">geno[,,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]</code></pre></div>
<pre><code>## , , 1
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    0
##   [2,]    0    0    0    0    0
## 
## , , 2
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    0
##   [2,]    0    0    0    0    0</code></pre>
<p>The <a href="http://bioconductor.org/packages/SeqVarTools">SeqVarTools package</a> has some additional functions for interacting with SeqArray-format GDS files.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SeqVarTools)

<span class="co"># return genotypes in matrix format</span>
<span class="kw">getGenotype</span>(gds)</code></pre></div>
<pre><code>##          variant
## sample    1     2     3     4     5     6     7     8     9     10   
##   HG00096 &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot;
##   HG00097 &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot;
##   HG00099 &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot;
##   HG00100 &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot;
##   HG00101 &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">getGenotypeAlleles</span>(gds)</code></pre></div>
<pre><code>##          variant
## sample    1     2     3     4     5     6     7     8     9     10   
##   HG00096 &quot;C|C&quot; &quot;C|C&quot; &quot;C|C&quot; &quot;C|C&quot; &quot;T|T&quot; &quot;G|G&quot; &quot;G|G&quot; &quot;A|A&quot; &quot;A|A&quot; &quot;C|C&quot;
##   HG00097 &quot;C|C&quot; &quot;C|C&quot; &quot;C|C&quot; &quot;C|C&quot; &quot;T|T&quot; &quot;G|G&quot; &quot;G|G&quot; &quot;A|A&quot; &quot;A|A&quot; &quot;C|C&quot;
##   HG00099 &quot;C|C&quot; &quot;C|C&quot; &quot;C|C&quot; &quot;C|C&quot; &quot;T|T&quot; &quot;G|G&quot; &quot;G|G&quot; &quot;A|A&quot; &quot;A|A&quot; &quot;C|C&quot;
##   HG00100 &quot;C|C&quot; &quot;C|C&quot; &quot;C|C&quot; &quot;C|C&quot; &quot;T|T&quot; &quot;G|G&quot; &quot;G|G&quot; &quot;A|A&quot; &quot;A|A&quot; &quot;C|C&quot;
##   HG00101 &quot;C|C&quot; &quot;C|C&quot; &quot;C|C&quot; &quot;C|C&quot; &quot;T|T&quot; &quot;G|G&quot; &quot;G|G&quot; &quot;A|A&quot; &quot;A|A&quot; &quot;C|C&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">refDosage</span>(gds)</code></pre></div>
<pre><code>##          variant
## sample    1 2 3 4 5 6 7 8 9 10
##   HG00096 2 2 2 2 2 2 2 2 2  2
##   HG00097 2 2 2 2 2 2 2 2 2  2
##   HG00099 2 2 2 2 2 2 2 2 2  2
##   HG00100 2 2 2 2 2 2 2 2 2  2
##   HG00101 2 2 2 2 2 2 2 2 2  2</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">altDosage</span>(gds)</code></pre></div>
<pre><code>##          variant
## sample    1 2 3 4 5 6 7 8 9 10
##   HG00096 0 0 0 0 0 0 0 0 0  0
##   HG00097 0 0 0 0 0 0 0 0 0  0
##   HG00099 0 0 0 0 0 0 0 0 0  0
##   HG00100 0 0 0 0 0 0 0 0 0  0
##   HG00101 0 0 0 0 0 0 0 0 0  0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># look at reference and alternate alleles</span>
<span class="kw">refChar</span>(gds)</code></pre></div>
<pre><code>##  [1] &quot;C&quot; &quot;C&quot; &quot;C&quot; &quot;C&quot; &quot;T&quot; &quot;G&quot; &quot;G&quot; &quot;A&quot; &quot;A&quot; &quot;C&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">altChar</span>(gds)</code></pre></div>
<pre><code>##  [1] &quot;G&quot; &quot;T&quot; &quot;T&quot; &quot;T&quot; &quot;C&quot; &quot;A&quot; &quot;A&quot; &quot;T&quot; &quot;C&quot; &quot;T&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># reset the filter to all variants and samples</span>
<span class="kw">seqResetFilter</span>(gds)</code></pre></div>
<pre><code>## # of selected samples: 1,126
## # of selected variants: 1,120</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># how many alleles for each variant?</span>
n &lt;-<span class="st"> </span><span class="kw">seqNumAllele</span>(gds)
<span class="kw">table</span>(n)</code></pre></div>
<pre><code>## n
##    2    3    4 
## 1099   20    1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># some variants have more than one alternate allele</span>
multi.allelic &lt;-<span class="st"> </span><span class="kw">which</span>(n <span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span>)
<span class="kw">altChar</span>(gds)[multi.allelic]</code></pre></div>
<pre><code>##  [1] &quot;GT,G&quot;            &quot;G,T&quot;             &quot;A,T&quot;            
##  [4] &quot;A,T&quot;             &quot;ATG,ATGTG&quot;       &quot;C,G&quot;            
##  [7] &quot;A,T&quot;             &quot;C,T&quot;             &quot;A,C&quot;            
## [10] &quot;TAA,T&quot;           &quot;GTTA,GTTT&quot;       &quot;GCC,GCCC,G&quot;     
## [13] &quot;A,C&quot;             &quot;A,C&quot;             &quot;A,C&quot;            
## [16] &quot;CAAGCAT,CGAGCAT&quot; &quot;CATTATT,C&quot;       &quot;AT,C&quot;           
## [19] &quot;TGTGA,C&quot;         &quot;CCATT,CCATTCATT&quot; &quot;C,G&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># extract a particular alternate allele</span>
<span class="co"># first alternate</span>
<span class="kw">altChar</span>(gds, <span class="dt">n=</span><span class="dv">1</span>)[multi.allelic]</code></pre></div>
<pre><code>##  [1] &quot;GT&quot;      &quot;G&quot;       &quot;A&quot;       &quot;A&quot;       &quot;ATG&quot;     &quot;C&quot;       &quot;A&quot;      
##  [8] &quot;C&quot;       &quot;A&quot;       &quot;TAA&quot;     &quot;GTTA&quot;    &quot;GCC&quot;     &quot;A&quot;       &quot;A&quot;      
## [15] &quot;A&quot;       &quot;CAAGCAT&quot; &quot;CATTATT&quot; &quot;AT&quot;      &quot;TGTGA&quot;   &quot;CCATT&quot;   &quot;C&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># second alternate</span>
<span class="kw">altChar</span>(gds, <span class="dt">n=</span><span class="dv">2</span>)[multi.allelic]</code></pre></div>
<pre><code>##  [1] &quot;G&quot;         &quot;T&quot;         &quot;T&quot;         &quot;T&quot;         &quot;ATGTG&quot;    
##  [6] &quot;G&quot;         &quot;T&quot;         &quot;T&quot;         &quot;C&quot;         &quot;T&quot;        
## [11] &quot;GTTT&quot;      &quot;GCCC&quot;      &quot;C&quot;         &quot;C&quot;         &quot;C&quot;        
## [16] &quot;CGAGCAT&quot;   &quot;C&quot;         &quot;C&quot;         &quot;C&quot;         &quot;CCATTCATT&quot;
## [21] &quot;G&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># how many variants are SNVs vs INDELs?</span>
<span class="kw">table</span>(<span class="kw">isSNV</span>(gds, <span class="dt">biallelic=</span><span class="ot">TRUE</span>))</code></pre></div>
<pre><code>## 
## FALSE  TRUE 
##   110  1010</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(<span class="kw">isSNV</span>(gds, <span class="dt">biallelic=</span><span class="ot">FALSE</span>))</code></pre></div>
<pre><code>## 
## FALSE  TRUE 
##    99  1021</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># 11 SNVs are multi-allelic</span></code></pre></div>
</div>
<div id="exercises" class="section level2">
<h2><span class="header-section-number">2.2</span> Exercises</h2>
<ol style="list-style-type: decimal">
<li>Set a filter selecting only multi-allelic variants. Inspect their genotypes using the different methods you learned above. Use the <code>alleleDosage</code> method to find dosage for the second (and third, etc.) alternate allele.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqSetFilter</span>(gds, <span class="dt">variant.sel=</span>multi.allelic)</code></pre></div>
<pre><code>## # of selected variants: 21</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">geno &lt;-<span class="st"> </span><span class="kw">seqGetData</span>(gds, <span class="st">&quot;genotype&quot;</span>)
<span class="kw">dim</span>(geno)</code></pre></div>
<pre><code>## [1]    2 1126   21</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">geno[,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,]</code></pre></div>
<pre><code>## , , 1
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    1    1    0    1    1
##   [2,]    0    1    1    1    1
## 
## , , 2
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    0
##   [2,]    0    0    0    0    0
## 
## , , 3
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    0
##   [2,]    0    0    0    0    0
## 
## , , 4
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    1    0    0    1    0
##   [2,]    0    0    0    0    1
## 
## , , 5
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    0
##   [2,]    0    0    0    0    0
## 
## , , 6
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    0
##   [2,]    0    0    0    0    0
## 
## , , 7
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    0
##   [2,]    0    0    0    0    0
## 
## , , 8
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    0
##   [2,]    0    0    0    0    0
## 
## , , 9
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    2
##   [2,]    0    0    0    0    0
## 
## , , 10
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    1    2    0    0    1
##   [2,]    0    2    0    2    1
## 
## , , 11
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    1    0    0
##   [2,]    1    1    0    1    0
## 
## , , 12
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    3    1    1    1    3
##   [2,]    3    3    3    1    0
## 
## , , 13
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    2    0
##   [2,]    0    0    0    0    0
## 
## , , 14
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    0
##   [2,]    0    0    0    0    0
## 
## , , 15
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    2    0    0    0    2
##   [2,]    0    0    0    2    0
## 
## , , 16
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    2    2    2    2    2
##   [2,]    2    2    2    2    2
## 
## , , 17
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    2    2    2    2    2
##   [2,]    2    2    2    2    2
## 
## , , 18
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    0
##   [2,]    0    0    0    0    0
## 
## , , 19
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    0
##   [2,]    0    0    0    0    0
## 
## , , 20
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    1
##   [2,]    0    0    0    0    0
## 
## , , 21
## 
##       sample
## allele [,1] [,2] [,3] [,4] [,5]
##   [1,]    0    0    0    0    0
##   [2,]    0    0    0    2    0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">geno &lt;-<span class="st"> </span><span class="kw">getGenotype</span>(gds)
<span class="kw">dim</span>(geno)</code></pre></div>
<pre><code>## [1] 1126   21</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(geno)</code></pre></div>
<pre><code>##          variant
## sample    30    69    73    161   162   195   243   253   407   431  
##   HG00096 &quot;1|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;1|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;1|0&quot;
##   HG00097 &quot;1|1&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;2|2&quot;
##   HG00099 &quot;0|1&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot;
##   HG00100 &quot;1|1&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;1|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|2&quot;
##   HG00101 &quot;1|1&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|1&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;2|0&quot; &quot;1|1&quot;
##   HG00102 &quot;0|1&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;1|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;2|2&quot;
##          variant
## sample    434   610   627   645   689   756   765   814   988   1014 
##   HG00096 &quot;0|1&quot; &quot;3|3&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;2|0&quot; &quot;2|2&quot; &quot;2|2&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot;
##   HG00097 &quot;0|1&quot; &quot;1|3&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;2|2&quot; &quot;2|2&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot;
##   HG00099 &quot;1|0&quot; &quot;1|3&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;2|2&quot; &quot;2|2&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot;
##   HG00100 &quot;0|1&quot; &quot;1|1&quot; &quot;2|0&quot; &quot;0|0&quot; &quot;0|2&quot; &quot;2|2&quot; &quot;2|2&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|0&quot;
##   HG00101 &quot;0|0&quot; &quot;3|0&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;2|0&quot; &quot;2|2&quot; &quot;2|2&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;1|0&quot;
##   HG00102 &quot;0|1&quot; &quot;3|3&quot; &quot;0|2&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;2|2&quot; &quot;1|2&quot; &quot;0|0&quot; &quot;0|0&quot; &quot;0|1&quot;
##          variant
## sample    1056 
##   HG00096 &quot;0|0&quot;
##   HG00097 &quot;0|0&quot;
##   HG00099 &quot;0|0&quot;
##   HG00100 &quot;0|2&quot;
##   HG00101 &quot;0|0&quot;
##   HG00102 &quot;0|2&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">geno &lt;-<span class="st"> </span><span class="kw">getGenotypeAlleles</span>(gds)
<span class="kw">head</span>(geno)</code></pre></div>
<pre><code>##          variant
## sample    30       69    73    161   162   195   243   253   407  
##   HG00096 &quot;GT|GTT&quot; &quot;A|A&quot; &quot;G|G&quot; &quot;A|G&quot; &quot;A|A&quot; &quot;A|A&quot; &quot;C|C&quot; &quot;G|G&quot; &quot;T|T&quot;
##   HG00097 &quot;GT|GT&quot;  &quot;A|A&quot; &quot;G|G&quot; &quot;G|G&quot; &quot;A|A&quot; &quot;A|A&quot; &quot;C|C&quot; &quot;G|G&quot; &quot;T|T&quot;
##   HG00099 &quot;GTT|GT&quot; &quot;A|A&quot; &quot;G|G&quot; &quot;G|G&quot; &quot;A|A&quot; &quot;A|A&quot; &quot;C|C&quot; &quot;G|G&quot; &quot;T|T&quot;
##   HG00100 &quot;GT|GT&quot;  &quot;A|A&quot; &quot;G|G&quot; &quot;A|G&quot; &quot;A|A&quot; &quot;A|A&quot; &quot;C|C&quot; &quot;G|G&quot; &quot;T|T&quot;
##   HG00101 &quot;GT|GT&quot;  &quot;A|A&quot; &quot;G|G&quot; &quot;G|A&quot; &quot;A|A&quot; &quot;A|A&quot; &quot;C|C&quot; &quot;G|G&quot; &quot;C|T&quot;
##   HG00102 &quot;GTT|GT&quot; &quot;A|A&quot; &quot;G|G&quot; &quot;A|G&quot; &quot;A|A&quot; &quot;A|A&quot; &quot;C|C&quot; &quot;G|G&quot; &quot;T|T&quot;
##          variant
## sample    431       434      610       627   645   689   756              
##   HG00096 &quot;TAA|TA&quot;  &quot;G|GTTA&quot; &quot;G|G&quot;     &quot;G|G&quot; &quot;G|G&quot; &quot;C|G&quot; &quot;CGAGCAT|CGAGCAT&quot;
##   HG00097 &quot;T|T&quot;     &quot;G|GTTA&quot; &quot;GCC|G&quot;   &quot;G|G&quot; &quot;G|G&quot; &quot;G|G&quot; &quot;CGAGCAT|CGAGCAT&quot;
##   HG00099 &quot;TA|TA&quot;   &quot;GTTA|G&quot; &quot;GCC|G&quot;   &quot;G|G&quot; &quot;G|G&quot; &quot;G|G&quot; &quot;CGAGCAT|CGAGCAT&quot;
##   HG00100 &quot;TA|T&quot;    &quot;G|GTTA&quot; &quot;GCC|GCC&quot; &quot;C|G&quot; &quot;G|G&quot; &quot;G|C&quot; &quot;CGAGCAT|CGAGCAT&quot;
##   HG00101 &quot;TAA|TAA&quot; &quot;G|G&quot;    &quot;G|GC&quot;    &quot;G|G&quot; &quot;G|G&quot; &quot;C|G&quot; &quot;CGAGCAT|CGAGCAT&quot;
##   HG00102 &quot;T|T&quot;     &quot;G|GTTA&quot; &quot;G|G&quot;     &quot;G|C&quot; &quot;G|G&quot; &quot;G|G&quot; &quot;CGAGCAT|CGAGCAT&quot;
##          variant
## sample    765         814     988           1014      1056 
##   HG00096 &quot;C|C&quot;       &quot;CT|CT&quot; &quot;CGTGA|CGTGA&quot; &quot;C|C&quot;     &quot;A|A&quot;
##   HG00097 &quot;C|C&quot;       &quot;CT|CT&quot; &quot;CGTGA|CGTGA&quot; &quot;C|C&quot;     &quot;A|A&quot;
##   HG00099 &quot;C|C&quot;       &quot;CT|CT&quot; &quot;CGTGA|CGTGA&quot; &quot;C|C&quot;     &quot;A|A&quot;
##   HG00100 &quot;C|C&quot;       &quot;CT|CT&quot; &quot;CGTGA|CGTGA&quot; &quot;C|C&quot;     &quot;A|G&quot;
##   HG00101 &quot;C|C&quot;       &quot;CT|CT&quot; &quot;CGTGA|CGTGA&quot; &quot;CCATT|C&quot; &quot;A|A&quot;
##   HG00102 &quot;CATTATT|C&quot; &quot;CT|CT&quot; &quot;CGTGA|CGTGA&quot; &quot;C|CCATT&quot; &quot;A|G&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dos &lt;-<span class="st"> </span><span class="kw">refDosage</span>(gds)
<span class="kw">head</span>(dos)</code></pre></div>
<pre><code>##          variant
## sample    30 69 73 161 162 195 243 253 407 431 434 610 627 645 689 756 765
##   HG00096  1  2  2   1   2   2   2   2   2   1   1   0   2   2   1   0   0
##   HG00097  0  2  2   2   2   2   2   2   2   0   1   0   2   2   2   0   0
##   HG00099  1  2  2   2   2   2   2   2   2   2   1   0   2   2   2   0   0
##   HG00100  0  2  2   1   2   2   2   2   2   1   1   0   1   2   1   0   0
##   HG00101  0  2  2   1   2   2   2   2   1   0   2   1   2   2   1   0   0
##   HG00102  1  2  2   1   2   2   2   2   2   0   1   0   1   2   2   0   0
##          variant
## sample    814 988 1014 1056
##   HG00096   2   2    2    2
##   HG00097   2   2    2    2
##   HG00099   2   2    2    2
##   HG00100   2   2    2    1
##   HG00101   2   2    1    2
##   HG00102   2   2    1    1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dos &lt;-<span class="st"> </span><span class="kw">altDosage</span>(gds)
<span class="kw">head</span>(dos)</code></pre></div>
<pre><code>##          variant
## sample    30 69 73 161 162 195 243 253 407 431 434 610 627 645 689 756 765
##   HG00096  1  0  0   1   0   0   0   0   0   1   1   2   0   0   1   2   2
##   HG00097  2  0  0   0   0   0   0   0   0   2   1   2   0   0   0   2   2
##   HG00099  1  0  0   0   0   0   0   0   0   0   1   2   0   0   0   2   2
##   HG00100  2  0  0   1   0   0   0   0   0   1   1   2   1   0   1   2   2
##   HG00101  2  0  0   1   0   0   0   0   1   2   0   1   0   0   1   2   2
##   HG00102  1  0  0   1   0   0   0   0   0   2   1   2   1   0   0   2   2
##          variant
## sample    814 988 1014 1056
##   HG00096   0   0    0    0
##   HG00097   0   0    0    0
##   HG00099   0   0    0    0
##   HG00100   0   0    0    1
##   HG00101   0   0    1    0
##   HG00102   0   0    1    1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dos &lt;-<span class="st"> </span><span class="kw">alleleDosage</span>(gds, <span class="dt">n=</span><span class="dv">2</span>)
<span class="kw">head</span>(dos)</code></pre></div>
<pre><code>##          variant
## sample    30 69 73 161 162 195 243 253 407 431 434 610 627 645 689 756 765
##   HG00096  0  0  0   0   0   0   0   0   0   0   0   0   0   0   1   2   2
##   HG00097  0  0  0   0   0   0   0   0   0   2   0   0   0   0   0   2   2
##   HG00099  0  0  0   0   0   0   0   0   0   0   0   0   0   0   0   2   2
##   HG00100  0  0  0   0   0   0   0   0   0   1   0   0   1   0   1   2   2
##   HG00101  0  0  0   0   0   0   0   0   1   0   0   0   0   0   1   2   2
##   HG00102  0  0  0   0   0   0   0   0   0   2   0   0   1   0   0   2   1
##          variant
## sample    814 988 1014 1056
##   HG00096   0   0    0    0
##   HG00097   0   0    0    0
##   HG00099   0   0    0    0
##   HG00100   0   0    0    1
##   HG00101   0   0    0    0
##   HG00102   0   0    0    1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dos &lt;-<span class="st"> </span><span class="kw">alleleDosage</span>(gds, <span class="dt">n=</span><span class="dv">3</span>)
<span class="kw">head</span>(dos)</code></pre></div>
<pre><code>##          variant
## sample    30 69 73 161 162 195 243 253 407 431 434 610 627 645 689 756 765
##   HG00096  0  0  0   0   0   0   0   0   0   0   0   2   0   0   0   0   0
##   HG00097  0  0  0   0   0   0   0   0   0   0   0   1   0   0   0   0   0
##   HG00099  0  0  0   0   0   0   0   0   0   0   0   1   0   0   0   0   0
##   HG00100  0  0  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
##   HG00101  0  0  0   0   0   0   0   0   0   0   0   1   0   0   0   0   0
##   HG00102  0  0  0   0   0   0   0   0   0   0   0   2   0   0   0   0   0
##          variant
## sample    814 988 1014 1056
##   HG00096   0   0    0    0
##   HG00097   0   0    0    0
##   HG00099   0   0    0    0
##   HG00100   0   0    0    0
##   HG00101   0   0    0    0
##   HG00102   0   0    0    0</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Use the <code>hwe</code> function in SeqVarTools to run a Hardy-Weinberg Equilibrium test on each variant. Identify a variant with low p-value and inspect its genotypes. (Note that the HWE test is only valid for biallelic variants, and will return <code>NA</code> for multiallelic variants.)</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqResetFilter</span>(gds)</code></pre></div>
<pre><code>## # of selected samples: 1,126
## # of selected variants: 1,120</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hwe.res &lt;-<span class="st"> </span><span class="kw">hwe</span>(gds)
lowp &lt;-<span class="st"> </span><span class="op">!</span><span class="kw">is.na</span>(hwe.res<span class="op">$</span>p) <span class="op">&amp;</span><span class="st"> </span>hwe.res<span class="op">$</span>p <span class="op">&lt;</span><span class="st"> </span><span class="fl">1e-4</span>
<span class="kw">head</span>(hwe.res[lowp,])</code></pre></div>
<pre><code>##     variant.id nAA nAa naa     afreq            p         f
## 75          75 702 336  88 0.7726465 1.070663e-06 0.1506466
## 92          92 632 381 113 0.7304618 3.558878e-06 0.1407120
## 98          98 672 335 119 0.7455595 2.369695e-12 0.2158342
## 105        105  93 272 761 0.2033748 7.851777e-16 0.2544970
## 114        114 299 482 345 0.4795737 1.745346e-06 0.1424409
## 150        150 471 447 208 0.6167851 8.020208e-08 0.1602251</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqSetFilter</span>(gds, <span class="dt">variant.id=</span><span class="dv">75</span>)</code></pre></div>
<pre><code>## # of selected variants: 1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(<span class="kw">getGenotype</span>(gds))</code></pre></div>
<pre><code>## 
## 0|0 0|1 1|0 1|1 
## 702 165 171  88</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(<span class="kw">refDosage</span>(gds))</code></pre></div>
<pre><code>## 
##   0   1   2 
##  88 336 702</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqClose</span>(gds)</code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="association-tests.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
