<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>5 PC-Relate | WISG Module 4: Computational Pipeline for WGS Data</title>
  <meta name="description" content="Course materials for WISG Module 4">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="5 PC-Relate | WISG Module 4: Computational Pipeline for WGS Data />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Course materials for WISG Module 4" />
  <meta name="github-repo" content="UW-GAC/WISG_2019" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="5 PC-Relate | WISG Module 4: Computational Pipeline for WGS Data />
  
  <meta name="twitter:description" content="Course materials for WISG Module 4" />
  



<meta name="date" content="2018-12-18">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="computing-a-grm.html">
<link rel="next" href="mixed-models.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#software-requirements"><i class="fa fa-check"></i><b>1.1</b> Software requirements</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#resources"><i class="fa fa-check"></i><b>1.2</b> Resources</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="gds-format.html"><a href="gds-format.html"><i class="fa fa-check"></i><b>2</b> GDS format</a><ul>
<li class="chapter" data-level="2.1" data-path="gds-format.html"><a href="gds-format.html#exploring-a-gds-file"><i class="fa fa-check"></i><b>2.1</b> Exploring a GDS file</a></li>
<li class="chapter" data-level="2.2" data-path="gds-format.html"><a href="gds-format.html#exercises"><i class="fa fa-check"></i><b>2.2</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="association-tests.html"><a href="association-tests.html"><i class="fa fa-check"></i><b>3</b> Association tests</a><ul>
<li class="chapter" data-level="3.1" data-path="association-tests.html"><a href="association-tests.html#null-model"><i class="fa fa-check"></i><b>3.1</b> Null model</a></li>
<li class="chapter" data-level="3.2" data-path="association-tests.html"><a href="association-tests.html#single-variant-tests"><i class="fa fa-check"></i><b>3.2</b> Single-variant tests</a></li>
<li class="chapter" data-level="3.3" data-path="association-tests.html"><a href="association-tests.html#exercises-1"><i class="fa fa-check"></i><b>3.3</b> Exercises</a></li>
<li class="chapter" data-level="3.4" data-path="association-tests.html"><a href="association-tests.html#sliding-window-tests"><i class="fa fa-check"></i><b>3.4</b> Sliding window tests</a><ul>
<li class="chapter" data-level="3.4.1" data-path="association-tests.html"><a href="association-tests.html#exercise"><i class="fa fa-check"></i><b>3.4.1</b> Exercise</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="computing-a-grm.html"><a href="computing-a-grm.html"><i class="fa fa-check"></i><b>4</b> Computing a GRM</a></li>
<li class="chapter" data-level="5" data-path="pc-relate.html"><a href="pc-relate.html"><i class="fa fa-check"></i><b>5</b> PC-Relate</a><ul>
<li class="chapter" data-level="5.1" data-path="pc-relate.html"><a href="pc-relate.html#king"><i class="fa fa-check"></i><b>5.1</b> KING</a></li>
<li class="chapter" data-level="5.2" data-path="pc-relate.html"><a href="pc-relate.html#pc-air"><i class="fa fa-check"></i><b>5.2</b> PC-AiR</a></li>
<li class="chapter" data-level="5.3" data-path="pc-relate.html"><a href="pc-relate.html#pc-relate-1"><i class="fa fa-check"></i><b>5.3</b> PC-Relate</a></li>
<li class="chapter" data-level="5.4" data-path="pc-relate.html"><a href="pc-relate.html#comparison-with-pedigree"><i class="fa fa-check"></i><b>5.4</b> Comparison with pedigree</a></li>
<li class="chapter" data-level="5.5" data-path="pc-relate.html"><a href="pc-relate.html#exercise-1"><i class="fa fa-check"></i><b>5.5</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="mixed-models.html"><a href="mixed-models.html"><i class="fa fa-check"></i><b>6</b> Mixed models</a><ul>
<li class="chapter" data-level="6.1" data-path="mixed-models.html"><a href="mixed-models.html#null-model-1"><i class="fa fa-check"></i><b>6.1</b> Null model</a></li>
<li class="chapter" data-level="6.2" data-path="mixed-models.html"><a href="mixed-models.html#single-variant-tests-1"><i class="fa fa-check"></i><b>6.2</b> Single-variant tests</a></li>
<li class="chapter" data-level="6.3" data-path="mixed-models.html"><a href="mixed-models.html#exercise-2"><i class="fa fa-check"></i><b>6.3</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="variant-annotation.html"><a href="variant-annotation.html"><i class="fa fa-check"></i><b>7</b> Variant annotation</a><ul>
<li class="chapter" data-level="7.1" data-path="variant-annotation.html"><a href="variant-annotation.html#using-bioconductor-annotation-resources"><i class="fa fa-check"></i><b>7.1</b> Using Bioconductor annotation resources</a></li>
<li class="chapter" data-level="7.2" data-path="variant-annotation.html"><a href="variant-annotation.html#aggregating-and-filtering-variants-using-annotation"><i class="fa fa-check"></i><b>7.2</b> Aggregating and filtering variants using annotation</a></li>
<li class="chapter" data-level="7.3" data-path="variant-annotation.html"><a href="variant-annotation.html#aggregate-unit-for-association-testing-exercise"><i class="fa fa-check"></i><b>7.3</b> Aggregate unit for association testing exercise</a></li>
<li class="chapter" data-level="7.4" data-path="variant-annotation.html"><a href="variant-annotation.html#association-testing-with-aggregate-units"><i class="fa fa-check"></i><b>7.4</b> Association testing with aggregate units</a></li>
<li class="chapter" data-level="7.5" data-path="variant-annotation.html"><a href="variant-annotation.html#exercise-3"><i class="fa fa-check"></i><b>7.5</b> Exercise</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">WISG Module 4: Computational Pipeline for WGS Data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="pc-relate" class="section level1">
<h1><span class="header-section-number">5</span> PC-Relate</h1>
<p>To disentangle ancestry from recent familial relatedness, we use the <a href="http://www.ncbi.nlm.nih.gov/pubmed/26748516">PC-Relate</a> method.</p>
<div id="king" class="section level2">
<h2><span class="header-section-number">5.1</span> KING</h2>
<p>Step 1 is to get initial estimates of kinship using <a href="http://www.ncbi.nlm.nih.gov/pubmed/20926424">KING</a>, which is robust to population structure but not admixture. The KING algorithm is available in SNPRelate. We select a subset of variants for this calculation with LD pruning.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># use a GDS file with all chromosomes</span>
<span class="kw">library</span>(SeqArray)
gdsfile &lt;-<span class="st"> &quot;data/1KG_phase3_subset.gds&quot;</span>
gds &lt;-<span class="st"> </span><span class="kw">seqOpen</span>(gdsfile)

<span class="co"># use a subset of 100 samples to make things run faster</span>
sampfile &lt;-<span class="st"> &quot;data/samples_subset100.RData&quot;</span>
sample.id &lt;-<span class="st"> </span>TopmedPipeline<span class="op">::</span><span class="kw">getobj</span>(sampfile)

<span class="co"># LD pruning to get variant set</span>
<span class="kw">library</span>(SNPRelate)
<span class="kw">set.seed</span>(<span class="dv">100</span>) <span class="co"># LD pruning has a random element; so make this reproducible</span>
snpset &lt;-<span class="st"> </span><span class="kw">snpgdsLDpruning</span>(gds, <span class="dt">sample.id=</span>sample.id, <span class="dt">method=</span><span class="st">&quot;corr&quot;</span>, 
                          <span class="dt">slide.max.bp=</span><span class="fl">10e6</span>, <span class="dt">ld.threshold=</span><span class="kw">sqrt</span>(<span class="fl">0.1</span>))</code></pre></div>
<pre><code>## SNV pruning based on LD:
## Excluding 1,120 SNVs on non-autosomes
## Calculating allele counts/frequencies ...
## 
[..................................................]  0%, ETC: ---    
[==================================================] 100%, completed in 0s
## Excluding 13,673 SNVs (monomorphic: TRUE, MAF: NaN, missing rate: NaN)
## Working space: 100 samples, 10,967 SNVs
##     using 1 (CPU) core
##     sliding window: 10,000,000 basepairs, Inf SNPs
##     |LD| threshold: 0.316228
##     method: correlation
## Chromosome 1: 31.25%, 350/1,120
## Chromosome 2: 31.61%, 354/1,120
## Chromosome 3: 31.07%, 348/1,120
## Chromosome 4: 31.25%, 350/1,120
## Chromosome 5: 29.82%, 334/1,120
## Chromosome 6: 31.25%, 350/1,120
## Chromosome 7: 28.39%, 318/1,120
## Chromosome 8: 26.16%, 293/1,120
## Chromosome 9: 28.39%, 318/1,120
## Chromosome 10: 28.30%, 317/1,120
## Chromosome 11: 26.88%, 301/1,120
## Chromosome 12: 28.57%, 320/1,120
## Chromosome 13: 25.27%, 283/1,120
## Chromosome 14: 24.29%, 272/1,120
## Chromosome 15: 22.59%, 253/1,120
## Chromosome 16: 22.41%, 251/1,120
## Chromosome 17: 22.32%, 250/1,120
## Chromosome 18: 23.75%, 266/1,120
## Chromosome 19: 21.96%, 246/1,120
## Chromosome 20: 20.80%, 233/1,120
## Chromosome 21: 17.95%, 201/1,120
## Chromosome 22: 17.41%, 195/1,120
## 6,403 markers are selected in total.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sapply</span>(snpset, length)</code></pre></div>
<pre><code>##  chr1  chr2  chr3  chr4  chr5  chr6  chr7  chr8  chr9 chr10 chr11 chr12 
##   350   354   348   350   334   350   318   293   318   317   301   320 
## chr13 chr14 chr15 chr16 chr17 chr18 chr19 chr20 chr21 chr22 
##   283   272   253   251   250   266   246   233   201   195</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pruned &lt;-<span class="st"> </span><span class="kw">unlist</span>(snpset, <span class="dt">use.names=</span><span class="ot">FALSE</span>)

<span class="co"># KING</span>
king &lt;-<span class="st"> </span><span class="kw">snpgdsIBDKING</span>(gds, <span class="dt">sample.id=</span>sample.id, <span class="dt">snp.id=</span>pruned)</code></pre></div>
<pre><code>## IBD analysis (KING method of moment) on genotypes:
## Calculating allele counts/frequencies ...
## 
[..................................................]  0%, ETC: ---    
[==================================================] 100%, completed in 0s
## Working space: 100 samples, 6,403 SNVs
##     using 1 (CPU) core
## No family is specified, and all individuals are treated as singletons.
## Relationship inference in the presence of population stratification.
## CPU capabilities: Double-Precision SSE2
## Tue Dec 18 19:39:12 2018    (internal increment: 65536)
## 
[..................................................]  0%, ETC: ---        
[==================================================] 100%, completed in 0s
## Tue Dec 18 19:39:12 2018    Done.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(king)</code></pre></div>
<pre><code>## [1] &quot;sample.id&quot; &quot;snp.id&quot;    &quot;afreq&quot;     &quot;IBS0&quot;      &quot;kinship&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(king<span class="op">$</span>kinship)</code></pre></div>
<pre><code>## [1] 100 100</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">kingMat &lt;-<span class="st"> </span>king<span class="op">$</span>kinship
<span class="kw">colnames</span>(kingMat) &lt;-<span class="st"> </span><span class="kw">rownames</span>(kingMat) &lt;-<span class="st"> </span>king<span class="op">$</span>sample.id</code></pre></div>
<p>We extract pairwise kinship estimates and IBS0 to plot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">kinship &lt;-<span class="st"> </span><span class="kw">snpgdsIBDSelection</span>(king)
<span class="kw">head</span>(kinship)</code></pre></div>
<pre><code>##       ID1     ID2       IBS0     kinship
## 1 HG00110 HG00116 0.02530064 -0.01532452
## 2 HG00110 HG00120 0.02748712 -0.02974760
## 3 HG00110 HG00128 0.02483211 -0.01852977
## 4 HG00110 HG00136 0.03014212 -0.05048077
## 5 HG00110 HG00137 0.02701859 -0.03846154
## 6 HG00110 HG00141 0.02920506 -0.04657452</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)
<span class="kw">ggplot</span>(kinship, <span class="kw">aes</span>(IBS0, kinship)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_hline</span>(<span class="dt">yintercept=</span><span class="dv">2</span><span class="op">^</span>(<span class="op">-</span><span class="kw">seq</span>(<span class="dv">3</span>,<span class="dv">9</span>,<span class="dv">2</span>)<span class="op">/</span><span class="dv">2</span>), <span class="dt">linetype=</span><span class="st">&quot;dashed&quot;</span>, <span class="dt">color=</span><span class="st">&quot;grey&quot;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">alpha=</span><span class="fl">0.5</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">ylab</span>(<span class="st">&quot;kinship estimate&quot;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme_bw</span>()</code></pre></div>
<p><img src="_main_files/figure-html/king_plot-1.png" width="672" /></p>
</div>
<div id="pc-air" class="section level2">
<h2><span class="header-section-number">5.2</span> PC-AiR</h2>
<p>The next step is <a href="http://www.ncbi.nlm.nih.gov/pubmed/25810074">PC-AiR</a>, in which we select a set of unrelated samples that is maximally informative about all ancestries in the sample. We use this unrelated set for Principal Component Analysis (PCA), then project the relatives onto the PCs.</p>
<p>First, we partition the samples into a related and unrelated set. We use a kinship threshold of degree 3, which corresponds to first cousins. This defines anyone less related than first cousins as “unrelated”. We load the GENESIS package. In the first iteration, we use the KING estimates for both kinship (<code>kinMat</code>) and ancestry divergence (<code>divMat</code>). KING kinship estimates are negative for samples with different ancestry.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(GENESIS)
sampset &lt;-<span class="st"> </span><span class="kw">pcairPartition</span>(<span class="dt">kinobj=</span>kingMat, <span class="dt">kin.thresh=</span><span class="dv">2</span><span class="op">^</span>(<span class="op">-</span><span class="dv">9</span><span class="op">/</span><span class="dv">2</span>),
                          <span class="dt">divobj=</span>kingMat, <span class="dt">div.thresh=</span><span class="op">-</span><span class="dv">2</span><span class="op">^</span>(<span class="op">-</span><span class="dv">9</span><span class="op">/</span><span class="dv">2</span>))
<span class="kw">names</span>(sampset)</code></pre></div>
<pre><code>## [1] &quot;rels&quot;   &quot;unrels&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sapply</span>(sampset, length)</code></pre></div>
<pre><code>##   rels unrels 
##     14     86</code></pre>
<p>Using the SNPRelate package, we run PCA on the unrelated set and project values for the related set. We use the same LD pruned set of variants again.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># run PCA on unrelated set</span>
pca.unrel &lt;-<span class="st"> </span><span class="kw">snpgdsPCA</span>(gds, <span class="dt">sample.id=</span>sampset<span class="op">$</span>unrels, <span class="dt">snp.id=</span>pruned)</code></pre></div>
<pre><code>## Principal Component Analysis (PCA) on genotypes:
## Calculating allele counts/frequencies ...
## 
[..................................................]  0%, ETC: ---    
[==================================================] 100%, completed in 0s
## Excluding 222 SNVs (monomorphic: TRUE, MAF: NaN, missing rate: NaN)
## Working space: 86 samples, 6,181 SNVs
##     using 1 (CPU) core
## CPU capabilities: Double-Precision SSE2
## Tue Dec 18 19:39:13 2018    (internal increment: 8712)
## 
[..................................................]  0%, ETC: ---        
[==================================================] 100%, completed in 0s
## Tue Dec 18 19:39:13 2018    Begin (eigenvalues and eigenvectors)
## Tue Dec 18 19:39:13 2018    Done.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># project values for relatives</span>
snp.load &lt;-<span class="st"> </span><span class="kw">snpgdsPCASNPLoading</span>(pca.unrel, <span class="dt">gdsobj=</span>gds)</code></pre></div>
<pre><code>## SNP loading:
## Working space: 86 samples, 6181 SNPs
##     using 1 (CPU) core
##     using the top 32 eigenvectors
## Tue Dec 18 19:39:13 2018    (internal increment: 65536)
## 
[..................................................]  0%, ETC: ---        
[==================================================] 100%, completed in 1s
## Tue Dec 18 19:39:14 2018    Done.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">samp.load &lt;-<span class="st"> </span><span class="kw">snpgdsPCASampLoading</span>(snp.load, <span class="dt">gdsobj=</span>gds, <span class="dt">sample.id=</span>sampset<span class="op">$</span>rels)</code></pre></div>
<pre><code>## Sample loading:
## Working space: 14 samples, 6181 SNPs
##     using 1 (CPU) core
##     using the top 32 eigenvectors
## Tue Dec 18 19:39:14 2018    (internal increment: 65536)
## 
[..................................................]  0%, ETC: ---        
[==================================================] 100%, completed in 0s
## Tue Dec 18 19:39:14 2018    Done.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># combine unrelated and related PCs and order as in GDS file</span>
pcs &lt;-<span class="st"> </span><span class="kw">rbind</span>(pca.unrel<span class="op">$</span>eigenvect, samp.load<span class="op">$</span>eigenvect)
<span class="kw">rownames</span>(pcs) &lt;-<span class="st"> </span><span class="kw">c</span>(pca.unrel<span class="op">$</span>sample.id, samp.load<span class="op">$</span>sample.id)
samp.ord &lt;-<span class="st"> </span><span class="kw">match</span>(sample.id, <span class="kw">rownames</span>(pcs))
pcs &lt;-<span class="st"> </span>pcs[samp.ord,]</code></pre></div>
<p>We need to determine which PCs are ancestry informative. To do this we need population information for the 1000 Genomes samples. This information is stored in an <code>AnnotatedDataFrame</code>, which is a data.frame with optional metadata describing the columns. The class is defined in the Biobase package. We load the stored object using the <code>getobj</code> function from the TopmedPipeline package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(Biobase)
sampfile &lt;-<span class="st"> &quot;data/sample_annotation.RData&quot;</span>
annot &lt;-<span class="st"> </span>TopmedPipeline<span class="op">::</span><span class="kw">getobj</span>(sampfile)
annot</code></pre></div>
<pre><code>## An object of class &#39;AnnotatedDataFrame&#39;
##   rowNames: 1 2 ... 2504 (1126 total)
##   varLabels: sample.id subject.id ... status (6 total)
##   varMetadata: labelDescription</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">pData</span>(annot))</code></pre></div>
<pre><code>##   sample.id subject.id Population          Population.Description sex
## 1   HG00096    HG00096        GBR British in England and Scotland   M
## 2   HG00097    HG00097        GBR British in England and Scotland   F
## 3   HG00099    HG00099        GBR British in England and Scotland   F
## 4   HG00100    HG00100        GBR British in England and Scotland   F
## 5   HG00101    HG00101        GBR British in England and Scotland   M
## 6   HG00102    HG00102        GBR British in England and Scotland   F
##   status
## 1      0
## 2      1
## 3      0
## 4      1
## 5      0
## 6      0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">varMetadata</span>(annot)</code></pre></div>
<pre><code>##                                     labelDescription
## sample.id                          sample identifier
## subject.id                        subject identifier
## Population                   population abbreviation
## Population.Description        population description
## sex                                              sex
## status                 simulated case/control status</code></pre>
<p>We make a parallel coordinates plot, color-coding by 1000 Genomes population. We load the <a href="http://dplyr.tidyverse.org">dplyr</a> package for data.frame manipulation.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pc.df &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(pcs)
<span class="kw">names</span>(pc.df) &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(pcs)
pc.df<span class="op">$</span>sample.id &lt;-<span class="st"> </span><span class="kw">row.names</span>(pcs)

<span class="kw">library</span>(dplyr)
annot &lt;-<span class="st"> </span><span class="kw">pData</span>(annot) <span class="op">%&gt;%</span>
<span class="st">        </span>dplyr<span class="op">::</span><span class="kw">select</span>(sample.id, Population)
pc.df &lt;-<span class="st"> </span><span class="kw">left_join</span>(pc.df, annot, <span class="dt">by=</span><span class="st">&quot;sample.id&quot;</span>)

<span class="kw">library</span>(GGally)
<span class="kw">library</span>(RColorBrewer)
pop.cols &lt;-<span class="st"> </span><span class="kw">setNames</span>(<span class="kw">brewer.pal</span>(<span class="dv">12</span>, <span class="st">&quot;Paired&quot;</span>),
                 <span class="kw">c</span>(<span class="st">&quot;ACB&quot;</span>, <span class="st">&quot;ASW&quot;</span>, <span class="st">&quot;CEU&quot;</span>, <span class="st">&quot;GBR&quot;</span>, <span class="st">&quot;CHB&quot;</span>, <span class="st">&quot;JPT&quot;</span>, <span class="st">&quot;CLM&quot;</span>, <span class="st">&quot;MXL&quot;</span>, <span class="st">&quot;LWK&quot;</span>, <span class="st">&quot;YRI&quot;</span>, <span class="st">&quot;GIH&quot;</span>, <span class="st">&quot;PUR&quot;</span>))
<span class="kw">ggparcoord</span>(pc.df, <span class="dt">columns=</span><span class="dv">1</span><span class="op">:</span><span class="dv">12</span>, <span class="dt">groupColumn=</span><span class="st">&quot;Population&quot;</span>, <span class="dt">scale=</span><span class="st">&quot;uniminmax&quot;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_color_manual</span>(<span class="dt">values=</span>pop.cols) <span class="op">+</span>
<span class="st">    </span><span class="kw">xlab</span>(<span class="st">&quot;PC&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">ylab</span>(<span class="st">&quot;&quot;</span>)</code></pre></div>
<p><img src="_main_files/figure-html/pcair_parcoord-1.png" width="672" /></p>
</div>
<div id="pc-relate-1" class="section level2">
<h2><span class="header-section-number">5.3</span> PC-Relate</h2>
<p>The first 2 PCs separate populations, so we use them to compute kinship estimates adjusting for ancestry. The PC-Relate function expects a <code>SeqVarData</code> object, which allows linking sample and variant annotation with a GDS file in a single object. We will cover these in more detail later for association testing, but for now we create a bare object with no annotation.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqResetFilter</span>(gds, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
<span class="kw">library</span>(SeqVarTools)
seqData &lt;-<span class="st"> </span><span class="kw">SeqVarData</span>(gds)
<span class="kw">seqSetFilter</span>(seqData, <span class="dt">variant.id=</span>pruned)</code></pre></div>
<pre><code>## # of selected variants: 6,403</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">iterator &lt;-<span class="st"> </span><span class="kw">SeqVarBlockIterator</span>(seqData, <span class="dt">verbose=</span><span class="ot">FALSE</span>)

pcrel &lt;-<span class="st"> </span><span class="kw">pcrelate</span>(iterator, <span class="dt">pcs=</span>pcs[,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>], <span class="dt">training.set=</span>sampset<span class="op">$</span>unrels, 
                  <span class="dt">sample.include=</span>sample.id)
<span class="kw">names</span>(pcrel)</code></pre></div>
<pre><code>## [1] &quot;kinBtwn&quot; &quot;kinSelf&quot;</code></pre>
<p>PC-Relate is an iterative method. Now that we have ancestry-adjusted kinship estimates, we can use them to better adjust for ancestry in the PCs. This time we use the <code>pcair</code> function, which combines partitioning the sample set and running PCA in one step. First we need to make a kinship matrix from the PC-Relate results. The KING matrix is still used for ancestry divergence.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pcrelMat &lt;-<span class="st"> </span><span class="kw">pcrelateToMatrix</span>(pcrel, <span class="dt">scaleKin=</span><span class="dv">1</span>, <span class="dt">verbose=</span><span class="ot">FALSE</span>)

<span class="kw">seqResetFilter</span>(seqData, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
pca &lt;-<span class="st"> </span><span class="kw">pcair</span>(seqData,
             <span class="dt">kinobj=</span>pcrelMat, <span class="dt">kin.thresh=</span><span class="dv">2</span><span class="op">^</span>(<span class="op">-</span><span class="dv">9</span><span class="op">/</span><span class="dv">2</span>),
             <span class="dt">divobj=</span>kingMat, <span class="dt">div.thresh=</span><span class="op">-</span><span class="dv">2</span><span class="op">^</span>(<span class="op">-</span><span class="dv">9</span><span class="op">/</span><span class="dv">2</span>),
             <span class="dt">sample.include=</span>sample.id, <span class="dt">snp.include=</span>pruned,
             <span class="dt">verbose=</span><span class="ot">FALSE</span>)
<span class="kw">names</span>(pca)</code></pre></div>
<pre><code>##  [1] &quot;vectors&quot;    &quot;values&quot;     &quot;rels&quot;       &quot;unrels&quot;     &quot;kin.thresh&quot;
##  [6] &quot;div.thresh&quot; &quot;sample.id&quot;  &quot;nsamp&quot;      &quot;nsnps&quot;      &quot;varprop&quot;   
## [11] &quot;call&quot;       &quot;method&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pcs &lt;-<span class="st"> </span>pca<span class="op">$</span>vectors
pc.df &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(pcs)
<span class="kw">names</span>(pc.df) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;PC&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="kw">ncol</span>(pcs))
pc.df<span class="op">$</span>sample.id &lt;-<span class="st"> </span><span class="kw">row.names</span>(pcs)
pc.df &lt;-<span class="st"> </span><span class="kw">left_join</span>(pc.df, annot, <span class="dt">by=</span><span class="st">&quot;sample.id&quot;</span>)

<span class="kw">ggplot</span>(pc.df, <span class="kw">aes</span>(PC1, PC2, <span class="dt">color=</span>Population)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_color_manual</span>(<span class="dt">values=</span>pop.cols)</code></pre></div>
<p><img src="_main_files/figure-html/pcair_2-1.png" width="672" /></p>
<p>Now we use the revised PCs to compute new kinship estimates. One can run the iteration multiple times and check for conversion, but usually two rounds are sufficient.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqSetFilter</span>(seqData, <span class="dt">variant.id=</span>pruned)</code></pre></div>
<pre><code>## # of selected variants: 6,403</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">iterator &lt;-<span class="st"> </span><span class="kw">SeqVarBlockIterator</span>(seqData, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
pcrel &lt;-<span class="st"> </span><span class="kw">pcrelate</span>(iterator, <span class="dt">pcs=</span>pcs[,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>], <span class="dt">training.set=</span>pca<span class="op">$</span>unrels, 
                  <span class="dt">sample.include=</span>sample.id)
<span class="kw">save</span>(pcrel, <span class="dt">file=</span><span class="st">&quot;data/pcrelate_kinship.RData&quot;</span>)</code></pre></div>
<p>We plot the kinship estimates from PC-Relate, and notice that the values for less related pairs are much better behaved.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">kinship &lt;-<span class="st"> </span>pcrel<span class="op">$</span>kinBtwn

<span class="kw">ggplot</span>(kinship, <span class="kw">aes</span>(k0, kin)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_hline</span>(<span class="dt">yintercept=</span><span class="dv">2</span><span class="op">^</span>(<span class="op">-</span><span class="kw">seq</span>(<span class="dv">3</span>,<span class="dv">9</span>,<span class="dv">2</span>)<span class="op">/</span><span class="dv">2</span>), <span class="dt">linetype=</span><span class="st">&quot;dashed&quot;</span>, <span class="dt">color=</span><span class="st">&quot;grey&quot;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">alpha=</span><span class="fl">0.5</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">ylab</span>(<span class="st">&quot;kinship estimate&quot;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme_bw</span>()</code></pre></div>
<p><img src="_main_files/figure-html/pcrelate_plot-1.png" width="672" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqClose</span>(gds)</code></pre></div>
</div>
<div id="comparison-with-pedigree" class="section level2">
<h2><span class="header-section-number">5.4</span> Comparison with pedigree</h2>
<p>We can detect pedigree errors and sample identity problems by comparing the pedigree with empirical kinship estimates. We use a function from the GWASTools package, <code>pedigreePairwiseRelatedness</code>, to get expected pairwise relationships based on the pedigree.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pedfile &lt;-<span class="st"> &quot;data/pedigree.RData&quot;</span>
<span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(pedfile)) <span class="kw">download.file</span>(<span class="kw">file.path</span>(workshop.path, pedfile), pedfile)
ped &lt;-<span class="st"> </span>TopmedPipeline<span class="op">::</span><span class="kw">getobj</span>(pedfile)
<span class="kw">head</span>(ped)</code></pre></div>
<pre><code>##   family individ  father  mother sex
## 1   BB01 HG01879       0       0   M
## 2   BB01 HG01880       0       0   F
## 3   BB01 HG01881 HG01879 HG01880   F
## 4   BB02 HG01882       0       0   M
## 5   BB02 HG01883       0       0   F
## 6   BB02 HG01888 HG01882 HG01883   M</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pw &lt;-<span class="st"> </span>GWASTools<span class="op">::</span><span class="kw">pedigreePairwiseRelatedness</span>(ped)
<span class="kw">names</span>(pw)</code></pre></div>
<pre><code>## [1] &quot;inbred.fam&quot;  &quot;inbred.KC&quot;   &quot;relativeprs&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rel &lt;-<span class="st"> </span>pw<span class="op">$</span>relativeprs
<span class="kw">head</span>(rel)</code></pre></div>
<pre><code>##   Individ1 Individ2 relation kinship family
## 1  HG01879  HG01880        U    0.00   BB01
## 2  HG01879  HG01881       PO    0.25   BB01
## 3  HG01880  HG01881       PO    0.25   BB01
## 4  HG01882  HG01883        U    0.00   BB02
## 5  HG01882  HG01888       PO    0.25   BB02
## 6  HG01883  HG01888       PO    0.25   BB02</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(rel<span class="op">$</span>relation)</code></pre></div>
<pre><code>## 
##   Av   FS GpGc  HAv   HS   PO    U 
##    2    6   16    1    3  616  330</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">distinct</span>(rel, relation, kinship) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">arrange</span>(<span class="op">-</span>kinship)</code></pre></div>
<pre><code>##   relation kinship
## 1       PO  0.2500
## 2       FS  0.2500
## 3       HS  0.1250
## 4     GpGc  0.1250
## 5       Av  0.1250
## 6      HAv  0.0625
## 7        U  0.0000</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## assign degrees to expected relationship pairs
rel &lt;-<span class="st"> </span>rel <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">exp.rel=</span><span class="kw">ifelse</span>(kinship <span class="op">==</span><span class="st"> </span><span class="fl">0.125</span>, <span class="st">&quot;Deg2&quot;</span>, <span class="kw">ifelse</span>(kinship <span class="op">==</span><span class="st"> </span><span class="fl">0.0625</span>, <span class="st">&quot;Deg3&quot;</span>, relation)),
           <span class="dt">pair=</span>GWASTools<span class="op">::</span><span class="kw">pasteSorted</span>(Individ1, Individ2)) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">select</span>(pair, family, relation, exp.rel)

## assign degrees to observed relationship pairs
cut.dup &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">/</span>(<span class="dv">2</span><span class="op">^</span>(<span class="dv">3</span><span class="op">/</span><span class="dv">2</span>))
cut.deg1 &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">/</span>(<span class="dv">2</span><span class="op">^</span>(<span class="dv">5</span><span class="op">/</span><span class="dv">2</span>))
cut.deg2 &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">/</span>(<span class="dv">2</span><span class="op">^</span>(<span class="dv">7</span><span class="op">/</span><span class="dv">2</span>))
cut.deg3 &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">/</span>(<span class="dv">2</span><span class="op">^</span>(<span class="dv">9</span><span class="op">/</span><span class="dv">2</span>))
cut.k0 &lt;-<span class="st"> </span><span class="fl">0.1</span>
kinship &lt;-<span class="st"> </span>kinship <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">obs.rel=</span><span class="kw">ifelse</span>(kin <span class="op">&gt;</span><span class="st"> </span>cut.dup, <span class="st">&quot;Dup&quot;</span>,
                   <span class="kw">ifelse</span>(kin <span class="op">&gt;</span><span class="st"> </span>cut.deg1 <span class="op">&amp;</span><span class="st"> </span>k0 <span class="op">&lt;</span><span class="st"> </span>cut.k0, <span class="st">&quot;PO&quot;</span>,
                   <span class="kw">ifelse</span>(kin <span class="op">&gt;</span><span class="st"> </span>cut.deg1, <span class="st">&quot;FS&quot;</span>,
                   <span class="kw">ifelse</span>(kin <span class="op">&gt;</span><span class="st"> </span>cut.deg2, <span class="st">&quot;Deg2&quot;</span>,
                   <span class="kw">ifelse</span>(kin <span class="op">&gt;</span><span class="st"> </span>cut.deg3, <span class="st">&quot;Deg3&quot;</span>, <span class="st">&quot;U&quot;</span>))))))
<span class="kw">table</span>(kinship<span class="op">$</span>obs.rel)</code></pre></div>
<pre><code>## 
## Deg2 Deg3   FS   PO    U 
##    3    6    2    7 4932</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># merge observed and expected relationships</span>
kin.obs &lt;-<span class="st"> </span>kinship <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">select</span>(ID1, ID2, kin, k0, obs.rel) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">pair=</span>GWASTools<span class="op">::</span><span class="kw">pasteSorted</span>(ID1, ID2)) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">left_join</span>(rel, <span class="dt">by=</span><span class="st">&quot;pair&quot;</span>) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">select</span>(<span class="op">-</span>pair) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">exp.rel=</span><span class="kw">ifelse</span>(<span class="kw">is.na</span>(exp.rel), <span class="st">&quot;U&quot;</span>, exp.rel)) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">filter</span>(<span class="op">!</span>(exp.rel <span class="op">==</span><span class="st"> &quot;U&quot;</span> <span class="op">&amp;</span><span class="st"> </span>obs.rel <span class="op">==</span><span class="st"> &quot;U&quot;</span>))
<span class="kw">table</span>(kin.obs<span class="op">$</span>exp.rel, kin.obs<span class="op">$</span>obs.rel)</code></pre></div>
<pre><code>##    
##     Deg2 Deg3 FS PO
##   U    3    6  2  7</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(kin.obs, <span class="kw">aes</span>(k0, kin, <span class="dt">color=</span>obs.rel)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>()</code></pre></div>
<p><img src="_main_files/figure-html/pedigree-1.png" width="672" /></p>
<p>All the observed relationships were unexpected. These samples are from 1000 Genomes sequencing, and known relatives were excluded from the released data. Here we have detected some cryptic relatives that were not annotated in the pedigree.</p>
</div>
<div id="exercise-1" class="section level2">
<h2><span class="header-section-number">5.5</span> Exercise</h2>
<p>Complete one round of iteration using all samples from the test dataset and plot the results. Be sure to examine the parallel coordinates plot to determine the appropriate number of PCs to give as an argument to <code>pcrelate</code>.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="computing-a-grm.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="mixed-models.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
